{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMsW6BiVLSuNlp2LVJ6ClTi",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Kjain02/ML/blob/main/Bioinformatics%20Chatbot.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install langchain==0.0.187\n",
        "!pip install xmltodict==0.13.0\n",
        "\n",
        "\n",
        "!pip install requests==2.31.0\n",
        "!pip install beautifulsoup4==4.12.2"
      ],
      "metadata": {
        "id": "t5dcxz0JTR6m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install --upgrade langchain openai -q\n",
        "!pip install unstructured -q\n",
        "!pip install unstructured[local-inference] -q\n",
        "!pip install detectron2@git+https://github.com/facebookresearch/detectron2.git@v0.6#egg=detectron2 -q\n",
        "!apt-get install poppler-utils\n"
      ],
      "metadata": {
        "id": "aFes6Wl07f3W"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "os.environ[\"OPENAI_API_KEY\"]=\"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\""
      ],
      "metadata": {
        "id": "cxOM05Ff7woZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install openai"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Y_pF_hgl0l1x",
        "outputId": "31f8710b-8682-4b49-a51e-a96352881a90"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: openai in /usr/local/lib/python3.10/dist-packages (0.27.8)\n",
            "Requirement already satisfied: requests>=2.20 in /usr/local/lib/python3.10/dist-packages (from openai) (2.31.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from openai) (4.66.0)\n",
            "Requirement already satisfied: aiohttp in /usr/local/lib/python3.10/dist-packages (from openai) (3.8.5)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (3.2.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (3.4)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (2.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (2023.7.22)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (23.1.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (6.0.4)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (4.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.9.2)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.4.0)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.3.1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "stkF1h82_yQq"
      },
      "outputs": [],
      "source": [
        "sitemap='/content/sitemap.xml'"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open(sitemap, 'r') as f:\n",
        "\tfile = f.read()"
      ],
      "metadata": {
        "id": "PvVH5GPHSqgO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from bs4 import BeautifulSoup\n",
        "\n",
        "def parse_sitemap(path):\n",
        "    # response = requests.get(path)\n",
        "    soup = BeautifulSoup(file, \"xml\")\n",
        "    urls = [element.text for element in soup.find_all(\"loc\")]\n",
        "    return urls\n",
        "\n",
        "sites = parse_sitemap(sitemap)"
      ],
      "metadata": {
        "id": "EaKpv__OKVYI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sites"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gCpEGhnAKVWH",
        "outputId": "ecc2bbf3-504d-4c76-d71c-c64f3dcdb97e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['https://en.wikipedia.org/wiki/Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Talk:Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/File:En-Bioinformatics.ogg',\n",
              " 'https://en.wikipedia.org/wiki/Bioinformatics_(journal)',\n",
              " 'https://en.wikipedia.org/wiki/National_Center_for_Biotechnology_Information',\n",
              " 'https://en.wikipedia.org/wiki/Biological_computation',\n",
              " 'https://en.wikipedia.org/wiki/Genetic_algorithm',\n",
              " 'https://en.wikipedia.org/wiki/File:En-us-bioinformatics.ogg',\n",
              " 'https://en.wikipedia.org/wiki/Interdisciplinary',\n",
              " 'https://en.wikipedia.org/wiki/Science',\n",
              " 'https://en.wikipedia.org/wiki/Biology',\n",
              " 'https://en.wikipedia.org/wiki/Chemistry',\n",
              " 'https://en.wikipedia.org/wiki/Physics',\n",
              " 'https://en.wikipedia.org/wiki/Computer_science',\n",
              " 'https://en.wikipedia.org/wiki/Computer_programming',\n",
              " 'https://en.wikipedia.org/wiki/Information_engineering_(field)',\n",
              " 'https://en.wikipedia.org/wiki/Mathematics',\n",
              " 'https://en.wikipedia.org/wiki/Statistics',\n",
              " 'https://en.wikipedia.org/wiki/Biological_data',\n",
              " 'https://en.wikipedia.org/wiki/Computational_biology',\n",
              " 'https://en.wikipedia.org/wiki/In_silico',\n",
              " 'https://en.wikipedia.org/wiki/Genomics',\n",
              " 'https://en.wikipedia.org/wiki/Gene',\n",
              " 'https://en.wikipedia.org/wiki/Nucleotide',\n",
              " 'https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism',\n",
              " 'https://en.wikipedia.org/wiki/Proteomics',\n",
              " 'https://en.wikipedia.org/wiki/Nucleic_acid',\n",
              " 'https://en.wikipedia.org/wiki/Protein',\n",
              " 'https://en.wikipedia.org/wiki/Signal_processing',\n",
              " 'https://en.wikipedia.org/wiki/Mutation',\n",
              " 'https://en.wikipedia.org/wiki/Text_mining',\n",
              " 'https://en.wikipedia.org/wiki/Ontology_(information_science)',\n",
              " 'https://en.wikipedia.org/wiki/Systems_biology',\n",
              " 'https://en.wikipedia.org/wiki/Structural_biology',\n",
              " 'https://en.wikipedia.org/wiki/Paulien_Hogeweg',\n",
              " 'https://en.wikipedia.org/wiki/Biochemistry',\n",
              " 'https://en.wikipedia.org/wiki/Human_Genome_Project',\n",
              " 'https://en.wikipedia.org/wiki/Algorithm',\n",
              " 'https://en.wikipedia.org/wiki/Graph_theory',\n",
              " 'https://en.wikipedia.org/wiki/Artificial_intelligence',\n",
              " 'https://en.wikipedia.org/wiki/Soft_computing',\n",
              " 'https://en.wikipedia.org/wiki/Data_mining',\n",
              " 'https://en.wikipedia.org/wiki/Image_processing',\n",
              " 'https://en.wikipedia.org/wiki/Computer_simulation',\n",
              " 'https://en.wikipedia.org/wiki/Discrete_mathematics',\n",
              " 'https://en.wikipedia.org/wiki/Control_theory',\n",
              " 'https://en.wikipedia.org/wiki/System_theory',\n",
              " 'https://en.wikipedia.org/wiki/Information_theory',\n",
              " 'https://en.wikipedia.org/wiki/Sequence',\n",
              " 'https://en.wikipedia.org/wiki/Protein_sequences',\n",
              " 'https://en.wikipedia.org/wiki/Frederick_Sanger',\n",
              " 'https://en.wikipedia.org/wiki/Insulin',\n",
              " 'https://en.wikipedia.org/wiki/Margaret_Oakley_Dayhoff',\n",
              " 'https://en.wikipedia.org/wiki/Molecular_evolution',\n",
              " 'https://en.wikipedia.org/wiki/Elvin_A._Kabat',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Dates_and_numbers',\n",
              " 'https://en.wikipedia.org/wiki/Amino_acid_sequence',\n",
              " 'https://en.wikipedia.org/wiki/Protein_domain',\n",
              " 'https://en.wikipedia.org/wiki/Protein_structure',\n",
              " 'https://en.wikipedia.org/wiki/Cluster_analysis',\n",
              " 'https://en.wikipedia.org/wiki/Pattern_recognition',\n",
              " 'https://en.wikipedia.org/wiki/Machine_learning',\n",
              " 'https://en.wikipedia.org/wiki/Biological_Data_Visualization',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_alignment',\n",
              " 'https://en.wikipedia.org/wiki/Gene_finding',\n",
              " 'https://en.wikipedia.org/wiki/Genome_assembly',\n",
              " 'https://en.wikipedia.org/wiki/Drug_design',\n",
              " 'https://en.wikipedia.org/wiki/Drug_discovery',\n",
              " 'https://en.wikipedia.org/wiki/Protein_structural_alignment',\n",
              " 'https://en.wikipedia.org/wiki/Protein_structure_prediction',\n",
              " 'https://en.wikipedia.org/wiki/Gene_expression',\n",
              " 'https://en.wikipedia.org/wiki/Protein%E2%80%93protein_interaction',\n",
              " 'https://en.wikipedia.org/wiki/Genome-wide_association_studies',\n",
              " 'https://en.wikipedia.org/wiki/Evolution',\n",
              " 'https://en.wikipedia.org/wiki/Cellular_model',\n",
              " 'https://en.wikipedia.org/wiki/Information_technologies',\n",
              " 'https://en.wikipedia.org/wiki/DNA',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_database',\n",
              " 'https://en.wikipedia.org/wiki/Alignment-free_sequence_analysis',\n",
              " 'https://en.wikipedia.org/wiki/Phi_X_174',\n",
              " 'https://en.wikipedia.org/wiki/Sequencing',\n",
              " 'https://en.wikipedia.org/wiki/DNA_sequence',\n",
              " 'https://en.wikipedia.org/wiki/Species',\n",
              " 'https://en.wikipedia.org/wiki/Molecular_systematics',\n",
              " 'https://en.wikipedia.org/wiki/Phylogenetic_tree',\n",
              " 'https://en.wikipedia.org/wiki/Computer_program',\n",
              " 'https://en.wikipedia.org/wiki/BLAST',\n",
              " 'https://en.wikipedia.org/wiki/DNA_sequencing',\n",
              " 'https://en.wikipedia.org/wiki/Base_calling',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_assembly',\n",
              " 'https://en.wikipedia.org/wiki/Shotgun_sequencing',\n",
              " 'https://en.wikipedia.org/wiki/The_Institute_for_Genomic_Research',\n",
              " 'https://en.wikipedia.org/wiki/Haemophilus_influenzae',\n",
              " 'https://en.wikipedia.org/wiki/Human_genome',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_analysis',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_mining',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_profiling_tool',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_motif',\n",
              " 'https://en.wikipedia.org/wiki/Gene_prediction',\n",
              " 'https://en.wikipedia.org/wiki/Genome_project',\n",
              " 'https://en.wikipedia.org/wiki/Ab_initio',\n",
              " 'https://en.wikipedia.org/wiki/Free-living',\n",
              " 'https://en.wikipedia.org/wiki/GeneMark',\n",
              " 'https://en.wikipedia.org/wiki/ENCODE',\n",
              " 'https://en.wikipedia.org/wiki/National_Human_Genome_Research_Institute',\n",
              " 'https://en.wikipedia.org/wiki/Homology_(biology)',\n",
              " 'https://en.wikipedia.org/wiki/Amino_acid',\n",
              " 'https://en.wikipedia.org/wiki/Transmembrane_domain',\n",
              " 'https://en.wikipedia.org/wiki/Computational_phylogenetics',\n",
              " 'https://en.wikipedia.org/wiki/Evolutionary_biology',\n",
              " 'https://en.wikipedia.org/wiki/Informatics_(academic_field)',\n",
              " 'https://en.wikipedia.org/wiki/Genomes',\n",
              " 'https://en.wikipedia.org/wiki/Gene_duplication',\n",
              " 'https://en.wikipedia.org/wiki/Horizontal_gene_transfer',\n",
              " 'https://en.wikipedia.org/wiki/Speciation',\n",
              " 'https://en.wikipedia.org/wiki/Population_genetics',\n",
              " 'https://en.wikipedia.org/wiki/Evolutionary_tree',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Words_to_watch',\n",
              " 'https://en.wikipedia.org/wiki/Comparative_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Genes',\n",
              " 'https://en.wikipedia.org/wiki/Endosymbiosis',\n",
              " 'https://en.wikipedia.org/wiki/Heuristics',\n",
              " 'https://en.wikipedia.org/wiki/Approximation_algorithms',\n",
              " 'https://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo',\n",
              " 'https://en.wikipedia.org/wiki/Bayesian_analysis',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_homology',\n",
              " 'https://en.wikipedia.org/wiki/Protein_family',\n",
              " 'https://en.wikipedia.org/wiki/Pan-genome',\n",
              " 'https://en.wikipedia.org/wiki/Monophyly',\n",
              " 'https://en.wikipedia.org/wiki/Mendelian_inheritance',\n",
              " 'https://en.wikipedia.org/wiki/Online_Mendelian_Inheritance_in_Man',\n",
              " 'https://en.wikipedia.org/wiki/Infertility',\n",
              " 'https://en.wikipedia.org/wiki/Breast_cancer',\n",
              " 'https://en.wikipedia.org/wiki/Alzheimer%27s_disease',\n",
              " 'https://en.wikipedia.org/wiki/Rare_functional_variant',\n",
              " 'https://en.wikipedia.org/wiki/Missing_heritability_problem',\n",
              " 'https://en.wikipedia.org/wiki/Whole_genome_sequencing',\n",
              " 'https://en.wikipedia.org/wiki/SNP_annotation',\n",
              " 'https://en.wikipedia.org/wiki/Oncogenomics',\n",
              " 'https://en.wikipedia.org/wiki/Cancer',\n",
              " 'https://en.wikipedia.org/wiki/Point_mutation',\n",
              " 'https://en.wikipedia.org/wiki/Oligonucleotide',\n",
              " 'https://en.wikipedia.org/wiki/Comparative_genomic_hybridization',\n",
              " 'https://en.wikipedia.org/wiki/Terabyte',\n",
              " 'https://en.wikipedia.org/wiki/Noise',\n",
              " 'https://en.wikipedia.org/wiki/Hidden_Markov_model',\n",
              " 'https://en.wikipedia.org/wiki/Copy_number_variation',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Citation_needed',\n",
              " 'https://en.wikipedia.org/wiki/Exome',\n",
              " 'https://en.wikipedia.org/wiki/Lesion',\n",
              " 'https://en.wikipedia.org/wiki/Messenger_RNA',\n",
              " 'https://en.wikipedia.org/wiki/DNA_microarray',\n",
              " 'https://en.wikipedia.org/wiki/Expressed_sequence_tag',\n",
              " 'https://en.wikipedia.org/wiki/Serial_analysis_of_gene_expression',\n",
              " 'https://en.wikipedia.org/wiki/Massively_parallel_signature_sequencing',\n",
              " 'https://en.wikipedia.org/wiki/RNA-Seq',\n",
              " 'https://en.wikipedia.org/wiki/Signal_(information_theory)',\n",
              " 'https://en.wikipedia.org/wiki/Epithelial',\n",
              " 'https://en.wikipedia.org/wiki/Protein_microarray',\n",
              " 'https://en.wikipedia.org/wiki/Mass_spectrometry',\n",
              " 'https://en.wikipedia.org/wiki/Immunohistochemistry',\n",
              " 'https://en.wikipedia.org/wiki/Tissue_microarray',\n",
              " 'https://en.wikipedia.org/wiki/Regulation_of_gene_expression',\n",
              " 'https://en.wikipedia.org/wiki/Hormone',\n",
              " 'https://en.wikipedia.org/wiki/Enhancer_(genetics)',\n",
              " 'https://en.wikipedia.org/wiki/Chromosome_conformation_capture',\n",
              " 'https://en.wikipedia.org/wiki/Microarray',\n",
              " 'https://en.wikipedia.org/wiki/Cell_cycle',\n",
              " 'https://en.wikipedia.org/wiki/Regulatory_elements',\n",
              " 'https://en.wikipedia.org/wiki/K-means_clustering',\n",
              " 'https://en.wikipedia.org/wiki/Self-organizing_map',\n",
              " 'https://en.wikipedia.org/wiki/Hierarchical_clustering',\n",
              " 'https://en.wikipedia.org/wiki/Consensus_clustering',\n",
              " 'https://en.wikipedia.org/wiki/Gene_ontology',\n",
              " 'https://en.wikipedia.org/wiki/Biological_database',\n",
              " 'https://en.wikipedia.org/wiki/Organelle',\n",
              " 'https://en.wikipedia.org/wiki/Protein_function_prediction',\n",
              " 'https://en.wikipedia.org/wiki/Cell_nucleus',\n",
              " 'https://en.wikipedia.org/wiki/RNA_splicing',\n",
              " 'https://en.wikipedia.org/wiki/Mitochondrion',\n",
              " 'https://en.wikipedia.org/wiki/Cellular_respiration',\n",
              " 'https://en.wikipedia.org/wiki/Metabolism',\n",
              " 'https://en.wikipedia.org/wiki/Protein_subcellular_localization_prediction',\n",
              " 'https://en.wikipedia.org/wiki/Nuclear_organization',\n",
              " 'https://en.wikipedia.org/wiki/Hi-C_(genomic_analysis_technique)',\n",
              " 'https://en.wikipedia.org/wiki/ChIA-PET',\n",
              " 'https://en.wikipedia.org/wiki/Chromatin',\n",
              " 'https://en.wikipedia.org/wiki/Topologically_Associating_Domain',\n",
              " 'https://en.wikipedia.org/wiki/Structural_bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Structural_motif',\n",
              " 'https://en.wikipedia.org/wiki/Structural_domain',\n",
              " 'https://en.wikipedia.org/wiki/Primary_structure',\n",
              " 'https://en.wikipedia.org/wiki/Codons',\n",
              " 'https://en.wikipedia.org/wiki/Prion',\n",
              " 'https://en.wikipedia.org/wiki/Bovine_spongiform_encephalopathy',\n",
              " 'https://en.wikipedia.org/wiki/Secondary_structure',\n",
              " 'https://en.wikipedia.org/wiki/Tertiary_structure',\n",
              " 'https://en.wikipedia.org/wiki/Quaternary_structure',\n",
              " 'https://en.wikipedia.org/wiki/Homology_modeling',\n",
              " 'https://en.wikipedia.org/wiki/Leghemoglobin',\n",
              " 'https://en.wikipedia.org/wiki/Protein_superfamily',\n",
              " 'https://en.wikipedia.org/wiki/Virtual_screening',\n",
              " 'https://en.wikipedia.org/wiki/QSAR',\n",
              " 'https://en.wikipedia.org/wiki/Deep-learning',\n",
              " 'https://en.wikipedia.org/wiki/AlphaFold',\n",
              " 'https://en.wikipedia.org/wiki/DeepMind',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Please_clarify',\n",
              " 'https://en.wikipedia.org/wiki/Computational_systems_biology',\n",
              " 'https://en.wikipedia.org/wiki/Biological_network',\n",
              " 'https://en.wikipedia.org/wiki/Interactome',\n",
              " 'https://en.wikipedia.org/wiki/Metabolic_network',\n",
              " 'https://en.wikipedia.org/wiki/Cell_(biology)',\n",
              " 'https://en.wikipedia.org/wiki/Enzyme',\n",
              " 'https://en.wikipedia.org/wiki/Signal_transduction',\n",
              " 'https://en.wikipedia.org/wiki/Gene_regulatory_network',\n",
              " 'https://en.wikipedia.org/wiki/Artificial_life',\n",
              " 'https://en.wikipedia.org/wiki/Treponema_pallidum',\n",
              " 'https://en.wikipedia.org/wiki/Syphilis',\n",
              " 'https://en.wikipedia.org/wiki/Protein%E2%80%93protein_interaction_prediction',\n",
              " 'https://en.wikipedia.org/wiki/X-ray_crystallography',\n",
              " 'https://en.wikipedia.org/wiki/Protein_nuclear_magnetic_resonance_spectroscopy',\n",
              " 'https://en.wikipedia.org/wiki/Protein%E2%80%93protein_docking',\n",
              " 'https://en.wikipedia.org/wiki/Biomedical_text_mining',\n",
              " 'https://en.wikipedia.org/wiki/Named-entity_recognition',\n",
              " 'https://en.wikipedia.org/wiki/Computational_linguistics',\n",
              " 'https://en.wikipedia.org/wiki/Medical_imaging',\n",
              " 'https://en.wikipedia.org/wiki/Image_analysis',\n",
              " 'https://en.wikipedia.org/wiki/Accuracy',\n",
              " 'https://en.wikipedia.org/wiki/Objectivity_(science)',\n",
              " 'https://en.wikipedia.org/wiki/Diagnostics',\n",
              " 'https://en.wikipedia.org/wiki/High-content_screening',\n",
              " 'https://en.wikipedia.org/wiki/Bioimage_informatics',\n",
              " 'https://en.wikipedia.org/wiki/Morphometrics',\n",
              " 'https://en.wikipedia.org/wiki/Gene_mapping',\n",
              " 'https://en.wikipedia.org/wiki/Sulston_score',\n",
              " 'https://en.wikipedia.org/wiki/Flow_cytometry_bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Flow_cytometry',\n",
              " 'https://en.wikipedia.org/wiki/Biodiversity_informatics',\n",
              " 'https://en.wikipedia.org/wiki/Biodiversity',\n",
              " 'https://en.wikipedia.org/wiki/Taxonomic_database',\n",
              " 'https://en.wikipedia.org/wiki/Microbiome',\n",
              " 'https://en.wikipedia.org/wiki/Phylogenetics',\n",
              " 'https://en.wikipedia.org/wiki/Niche_modelling',\n",
              " 'https://en.wikipedia.org/wiki/Species_richness',\n",
              " 'https://en.wikipedia.org/wiki/DNA_barcoding',\n",
              " 'https://en.wikipedia.org/wiki/Speciesism',\n",
              " 'https://en.wikipedia.org/wiki/Directed_acyclic_graph',\n",
              " 'https://en.wikipedia.org/wiki/Controlled_vocabulary',\n",
              " 'https://en.wikipedia.org/wiki/OBO_Foundry',\n",
              " 'https://en.wikipedia.org/wiki/List_of_biological_databases',\n",
              " 'https://en.wikipedia.org/wiki/Genbank',\n",
              " 'https://en.wikipedia.org/wiki/UniProt',\n",
              " 'https://en.wikipedia.org/wiki/Protein_Data_Bank',\n",
              " 'https://en.wikipedia.org/wiki/InterPro',\n",
              " 'https://en.wikipedia.org/wiki/Pfam',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_Read_Archive',\n",
              " 'https://en.wikipedia.org/wiki/KEGG',\n",
              " 'https://en.wikipedia.org/wiki/BioCyc_database_collection',\n",
              " 'https://en.wikipedia.org/wiki/GenoCAD',\n",
              " 'https://en.wikipedia.org/wiki/List_of_bioinformatics_software',\n",
              " 'https://en.wikipedia.org/wiki/List_of_bioinformatics_companies',\n",
              " 'https://en.wikipedia.org/wiki/Free_and_open-source_software',\n",
              " 'https://en.wikipedia.org/wiki/Open_code',\n",
              " 'https://en.wikipedia.org/wiki/Funding_of_science',\n",
              " 'https://en.wikipedia.org/wiki/Plug-in_(computing)',\n",
              " 'https://en.wikipedia.org/wiki/De_facto',\n",
              " 'https://en.wikipedia.org/wiki/Bioconductor',\n",
              " 'https://en.wikipedia.org/wiki/BioPerl',\n",
              " 'https://en.wikipedia.org/wiki/Biopython',\n",
              " 'https://en.wikipedia.org/wiki/BioJava',\n",
              " 'https://en.wikipedia.org/wiki/BioJS',\n",
              " 'https://en.wikipedia.org/wiki/BioRuby',\n",
              " 'https://en.wikipedia.org/wiki/Bioclipse',\n",
              " 'https://en.wikipedia.org/wiki/EMBOSS',\n",
              " 'https://en.wikipedia.org/wiki/.NET_Bio',\n",
              " 'https://en.wikipedia.org/wiki/Orange_(software)',\n",
              " 'https://en.wikipedia.org/wiki/Apache_Taverna',\n",
              " 'https://en.wikipedia.org/wiki/UGENE',\n",
              " 'https://en.wikipedia.org/wiki/Open_Bioinformatics_Foundation',\n",
              " 'https://en.wikipedia.org/wiki/Bioinformatics_Open_Source_Conference',\n",
              " 'https://en.wikipedia.org/wiki/SOAP',\n",
              " 'https://en.wikipedia.org/wiki/REST',\n",
              " 'https://en.wikipedia.org/wiki/European_Bioinformatics_Institute',\n",
              " 'https://en.wikipedia.org/wiki/Sequence_alignment_software',\n",
              " 'https://en.wikipedia.org/wiki/Multiple_sequence_alignment',\n",
              " 'https://en.wikipedia.org/wiki/Service-orientation',\n",
              " 'https://en.wikipedia.org/wiki/Bioinformatics_workflow_management_systems',\n",
              " 'https://en.wikipedia.org/wiki/Workflow_management_system',\n",
              " 'https://en.wikipedia.org/wiki/Provenance',\n",
              " 'https://en.wikipedia.org/wiki/Galaxy_(computational_biology)',\n",
              " 'https://en.wikipedia.org/wiki/Kepler_scientific_workflow_system',\n",
              " 'https://en.wikipedia.org/wiki/Anduril_(workflow_engine)',\n",
              " 'https://en.wikipedia.org/wiki/High-performance_Integrated_Virtual_Environment',\n",
              " 'https://en.wikipedia.org/wiki/Food_and_Drug_Administration',\n",
              " 'https://en.wikipedia.org/wiki/National_Institutes_of_Health',\n",
              " 'https://en.wikipedia.org/wiki/Human_Variome_Project',\n",
              " 'https://en.wikipedia.org/wiki/European_Federation_for_Medical_Informatics',\n",
              " 'https://en.wikipedia.org/wiki/Stanford_University',\n",
              " 'https://en.wikipedia.org/wiki/New_York_Genome_Center',\n",
              " 'https://en.wikipedia.org/wiki/George_Washington_University',\n",
              " 'https://en.wikipedia.org/wiki/BioCompute_Object',\n",
              " 'https://en.wikipedia.org/wiki/Master%27s_degree',\n",
              " 'https://en.wikipedia.org/wiki/Educational_technology',\n",
              " 'https://en.wikipedia.org/wiki/Rosalind_(education_platform)',\n",
              " 'https://en.wikipedia.org/wiki/Swiss_Institute_of_Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Canadian_Bioinformatics_Workshops',\n",
              " 'https://en.wikipedia.org/wiki/Creative_Commons',\n",
              " 'https://en.wikipedia.org/wiki/Raspberry_Pi',\n",
              " 'https://en.wikipedia.org/wiki/Massive_open_online_course',\n",
              " 'https://en.wikipedia.org/wiki/Coursera',\n",
              " 'https://en.wikipedia.org/wiki/University_of_California,_San_Diego',\n",
              " 'https://en.wikipedia.org/wiki/Johns_Hopkins_University',\n",
              " 'https://en.wikipedia.org/wiki/EdX',\n",
              " 'https://en.wikipedia.org/wiki/Harvard_University',\n",
              " 'https://en.wikipedia.org/wiki/Intelligent_Systems_for_Molecular_Biology',\n",
              " 'https://en.wikipedia.org/wiki/European_Conference_on_Computational_Biology',\n",
              " 'https://en.wikipedia.org/wiki/Research_in_Computational_Molecular_Biology',\n",
              " 'https://en.wikipedia.org/wiki/Bioinformatics_companies',\n",
              " 'https://en.wikipedia.org/wiki/Computational_biomodeling',\n",
              " 'https://en.wikipedia.org/wiki/Computational_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Cyberbiosecurity',\n",
              " 'https://en.wikipedia.org/wiki/Functional_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Health_informatics',\n",
              " 'https://en.wikipedia.org/wiki/International_Society_for_Computational_Biology',\n",
              " 'https://en.wikipedia.org/wiki/Jumping_library',\n",
              " 'https://en.wikipedia.org/wiki/List_of_bioinformatics_institutions',\n",
              " 'https://en.wikipedia.org/wiki/List_of_open-source_bioinformatics_software',\n",
              " 'https://en.wikipedia.org/wiki/List_of_bioinformatics_journals',\n",
              " 'https://en.wikipedia.org/wiki/Metabolomics',\n",
              " 'https://en.wikipedia.org/wiki/Nucleic_acid_sequence',\n",
              " 'https://en.wikipedia.org/wiki/Gene_Disease_Database',\n",
              " 'https://en.wikipedia.org/wiki/Doi_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/PMC_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/PMID_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/ISBN_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/Bibcode_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/ArXiv_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/S2CID_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/ISSN_(identifier)',\n",
              " 'https://en.wikipedia.org/wiki/Wayback_Machine',\n",
              " 'https://en.wikipedia.org/wiki/Srinivas_Aluru',\n",
              " 'https://en.wikipedia.org/wiki/Current_Protocols',\n",
              " 'https://en.wikipedia.org/wiki/Lior_Pachter',\n",
              " 'https://en.wikipedia.org/wiki/Bernd_Sturmfels',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:The_Wikipedia_Library',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Media_help',\n",
              " 'https://en.wikipedia.org/wiki/Wikipedia:Spoken_articles',\n",
              " 'https://en.wikipedia.org/wiki/File:Wiktionary-logo-en-v2.svg',\n",
              " 'https://en.wikipedia.org/wiki/File:Wikiversity_logo_2017.svg',\n",
              " 'https://en.wikipedia.org/wiki/File:Commons-logo.svg',\n",
              " 'https://en.wikipedia.org/wiki/Template:Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Template_talk:Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/GenBank',\n",
              " 'https://en.wikipedia.org/wiki/European_Nucleotide_Archive',\n",
              " 'https://en.wikipedia.org/wiki/DNA_Data_Bank_of_Japan',\n",
              " 'https://en.wikipedia.org/wiki/China_National_GeneBank',\n",
              " 'https://en.wikipedia.org/wiki/Protein_Information_Resource',\n",
              " 'https://en.wikipedia.org/wiki/BioNumbers',\n",
              " 'https://en.wikipedia.org/wiki/Ensembl_genome_database_project',\n",
              " 'https://en.wikipedia.org/wiki/Barcode_of_Life_Data_System',\n",
              " 'https://en.wikipedia.org/wiki/Saccharomyces_Genome_Database',\n",
              " 'https://en.wikipedia.org/wiki/FlyBase',\n",
              " 'https://en.wikipedia.org/wiki/VectorBase',\n",
              " 'https://en.wikipedia.org/wiki/WormBase',\n",
              " 'https://en.wikipedia.org/wiki/Rat_Genome_Database',\n",
              " 'https://en.wikipedia.org/wiki/PHI-base',\n",
              " 'https://en.wikipedia.org/wiki/The_Arabidopsis_Information_Resource',\n",
              " 'https://en.wikipedia.org/wiki/GISAID',\n",
              " 'https://en.wikipedia.org/wiki/Zebrafish_Information_Network',\n",
              " 'https://en.wikipedia.org/wiki/Bowtie_(sequence_analysis)',\n",
              " 'https://en.wikipedia.org/wiki/Clustal',\n",
              " 'https://en.wikipedia.org/wiki/HMMER',\n",
              " 'https://en.wikipedia.org/wiki/MUSCLE_(alignment_software)',\n",
              " 'https://en.wikipedia.org/wiki/Phylogenetic_Assignment_of_Named_Global_Outbreak_Lineages',\n",
              " 'https://en.wikipedia.org/wiki/SAMtools',\n",
              " 'https://en.wikipedia.org/wiki/Short_Oligonucleotide_Analysis_Package',\n",
              " 'https://en.wikipedia.org/wiki/TopHat_(bioinformatics)',\n",
              " 'https://en.wikipedia.org/wiki/ExPASy',\n",
              " 'https://en.wikipedia.org/wiki/Broad_Institute',\n",
              " 'https://en.wikipedia.org/wiki/Computational_Biology_Department',\n",
              " 'https://en.wikipedia.org/wiki/COSBI',\n",
              " 'https://en.wikipedia.org/wiki/Database_Center_for_Life_Science',\n",
              " 'https://en.wikipedia.org/wiki/European_Molecular_Biology_Laboratory',\n",
              " 'https://en.wikipedia.org/wiki/Flatiron_Institute',\n",
              " 'https://en.wikipedia.org/wiki/J._Craig_Venter_Institute',\n",
              " 'https://en.wikipedia.org/wiki/Max_Planck_Institute_of_Molecular_Cell_Biology_and_Genetics',\n",
              " 'https://en.wikipedia.org/wiki/National_Institute_of_Genetics',\n",
              " 'https://en.wikipedia.org/wiki/Netherlands_Bioinformatics_Centre',\n",
              " 'https://en.wikipedia.org/wiki/Philippine_Genome_Center',\n",
              " 'https://en.wikipedia.org/wiki/Scripps_Research',\n",
              " 'https://en.wikipedia.org/wiki/Wellcome_Sanger_Institute',\n",
              " 'https://en.wikipedia.org/wiki/Whitehead_Institute',\n",
              " 'https://en.wikipedia.org/wiki/African_Society_for_Bioinformatics_and_Computational_Biology',\n",
              " 'https://en.wikipedia.org/wiki/Australia_Bioinformatics_Resource',\n",
              " 'https://en.wikipedia.org/wiki/EMBnet',\n",
              " 'https://en.wikipedia.org/wiki/International_Nucleotide_Sequence_Database_Collaboration',\n",
              " 'https://en.wikipedia.org/wiki/International_Society_for_Biocuration',\n",
              " 'https://en.wikipedia.org/wiki/International_Society_for_Computational_Biology_Student_Council',\n",
              " 'https://en.wikipedia.org/wiki/Institute_of_Genomics_and_Integrative_Biology',\n",
              " 'https://en.wikipedia.org/wiki/Japanese_Society_for_Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Basel_Computational_Biology_Conference',\n",
              " 'https://en.wikipedia.org/wiki/International_Conference_on_Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/International_Conference_on_Computational_Intelligence_Methods_for_Bioinformatics_and_Biostatistics',\n",
              " 'https://en.wikipedia.org/wiki/ISCB_Africa_ASBCB_Conference_on_Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Pacific_Symposium_on_Biocomputing',\n",
              " 'https://en.wikipedia.org/wiki/CRAM_(file_format)',\n",
              " 'https://en.wikipedia.org/wiki/FASTA_format',\n",
              " 'https://en.wikipedia.org/wiki/FASTQ_format',\n",
              " 'https://en.wikipedia.org/wiki/NeXML_format',\n",
              " 'https://en.wikipedia.org/wiki/Nexus_file',\n",
              " 'https://en.wikipedia.org/wiki/Pileup_format',\n",
              " 'https://en.wikipedia.org/wiki/SAM_(file_format)',\n",
              " 'https://en.wikipedia.org/wiki/Stockholm_format',\n",
              " 'https://en.wikipedia.org/wiki/Variant_Call_Format',\n",
              " 'https://en.wikipedia.org/wiki/List_of_biobanks',\n",
              " 'https://en.wikipedia.org/wiki/Molecular_phylogenetics',\n",
              " 'https://en.wikipedia.org/wiki/Category:Bioinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Template:Informatics',\n",
              " 'https://en.wikipedia.org/wiki/Template_talk:Informatics',\n",
              " 'https://en.wikipedia.org/wiki/Information_processing',\n",
              " 'https://en.wikipedia.org/wiki/Perception',\n",
              " 'https://en.wikipedia.org/wiki/Attention',\n",
              " 'https://en.wikipedia.org/wiki/Social_influence',\n",
              " 'https://en.wikipedia.org/wiki/Remote_control',\n",
              " 'https://en.wikipedia.org/wiki/Communication',\n",
              " 'https://en.wikipedia.org/wiki/Reason',\n",
              " 'https://en.wikipedia.org/wiki/Learning',\n",
              " 'https://en.wikipedia.org/wiki/Information_storage',\n",
              " 'https://en.wikipedia.org/wiki/Decision-making',\n",
              " 'https://en.wikipedia.org/wiki/Event_processing',\n",
              " 'https://en.wikipedia.org/wiki/Semiotics',\n",
              " 'https://en.wikipedia.org/wiki/Signal',\n",
              " 'https://en.wikipedia.org/wiki/Data_processing',\n",
              " 'https://en.wikipedia.org/wiki/Stream_processing',\n",
              " 'https://en.wikipedia.org/wiki/Multi-agent_system',\n",
              " 'https://en.wikipedia.org/wiki/State_(computer_science)',\n",
              " 'https://en.wikipedia.org/wiki/Natural_computing',\n",
              " 'https://en.wikipedia.org/wiki/Information_processing_theory',\n",
              " 'https://en.wikipedia.org/wiki/Social_information_processing',\n",
              " 'https://en.wikipedia.org/wiki/Mixed_reality',\n",
              " 'https://en.wikipedia.org/wiki/Brain%E2%80%93computer_interface',\n",
              " 'https://en.wikipedia.org/wiki/Physical_computing',\n",
              " 'https://en.wikipedia.org/wiki/Human%E2%80%93computer_interaction',\n",
              " 'https://en.wikipedia.org/wiki/Processor_(computing)',\n",
              " 'https://en.wikipedia.org/wiki/Process_(computing)',\n",
              " 'https://en.wikipedia.org/wiki/Bio-inspired_computing',\n",
              " 'https://en.wikipedia.org/wiki/Ubiquitous_computing',\n",
              " 'https://en.wikipedia.org/wiki/Artificial_brain',\n",
              " 'https://en.wikipedia.org/wiki/Mind_uploading',\n",
              " 'https://en.wikipedia.org/wiki/Virtual_reality',\n",
              " 'https://en.wikipedia.org/wiki/Virtual_world',\n",
              " 'https://en.wikipedia.org/wiki/Genome_informatics',\n",
              " 'https://en.wikipedia.org/wiki/Cellular_computing',\n",
              " 'https://en.wikipedia.org/wiki/Computational_neuroscience',\n",
              " 'https://en.wikipedia.org/wiki/Neurocomputing',\n",
              " 'https://en.wikipedia.org/wiki/Mind',\n",
              " 'https://en.wikipedia.org/wiki/Intelligence',\n",
              " 'https://en.wikipedia.org/wiki/Cognitive_informatics',\n",
              " 'https://en.wikipedia.org/wiki/Neuroinformatics',\n",
              " 'https://en.wikipedia.org/wiki/Behavior_informatics',\n",
              " 'https://en.wikipedia.org/wiki/Neural_computation',\n",
              " 'https://en.wikipedia.org/wiki/Theory_of_computation',\n",
              " 'https://en.wikipedia.org/wiki/Algorithms',\n",
              " 'https://en.wikipedia.org/wiki/Information_structure',\n",
              " 'https://en.wikipedia.org/wiki/Circuit_(computer_science)',\n",
              " 'https://en.wikipedia.org/wiki/Computational_theory_of_mind',\n",
              " 'https://en.wikipedia.org/wiki/Philosophy_of_information',\n",
              " 'https://en.wikipedia.org/wiki/Philosophy_of_artificial_intelligence',\n",
              " 'https://en.wikipedia.org/wiki/Decision_theory',\n",
              " 'https://en.wikipedia.org/wiki/Systems_theory',\n",
              " 'https://en.wikipedia.org/wiki/Infosphere',\n",
              " 'https://en.wikipedia.org/wiki/Inforg',\n",
              " 'https://en.wikipedia.org/wiki/Decoding_the_Universe',\n",
              " 'https://en.wikipedia.org/wiki/Information_overload',\n",
              " 'https://en.wikipedia.org/wiki/Template:Genomics',\n",
              " 'https://en.wikipedia.org/wiki/Template_talk:Genomics',\n",
              " 'https://en.wikipedia.org/wiki/Omics',\n",
              " 'https://en.wikipedia.org/wiki/Cognitive_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Metagenomics',\n",
              " 'https://en.wikipedia.org/wiki/Human_Microbiome_Project',\n",
              " 'https://en.wikipedia.org/wiki/Pangenomics',\n",
              " 'https://en.wikipedia.org/wiki/Personal_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Population_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Social_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Structural_genomics',\n",
              " 'https://en.wikipedia.org/wiki/Biochip',\n",
              " 'https://en.wikipedia.org/wiki/Cheminformatics',\n",
              " 'https://en.wikipedia.org/wiki/Chemogenomics']"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "len(sites)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0neYfxHKKVUE",
        "outputId": "d9dde480-b392-4e5f-cf02-57e670aa1fdd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "488"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from langchain.document_loaders import UnstructuredURLLoader\n",
        "loader = UnstructuredURLLoader(urls=sites)\n",
        "documents = loader.load()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3vAophdYKVSL",
        "outputId": "22fda8d1-bc00-4fbc-847c-9f016b6bfbcd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n",
            "[nltk_data] Downloading package averaged_perceptron_tagger to\n",
            "[nltk_data]     /root/nltk_data...\n",
            "[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "documents[0]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HpCOXQnIKVQA",
        "outputId": "5258de57-667e-40bd-c744-30c6b4b79b8e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Document(page_content='Toggle the table of contents\\n\\nToggle the table of contents\\n\\nBioinformatics\\n\\n76 languages\\n\\nالعربية\\n\\nAzərbaycanca\\n\\nবাংলা\\n\\nBân-lâm-gú\\n\\nБеларуская\\n\\nБългарски\\n\\nBosanski\\n\\nCatalà\\n\\nČeština\\n\\nDansk\\n\\nDeutsch\\n\\nEesti\\n\\nΕλληνικά\\n\\nEspañol\\n\\nEsperanto\\n\\nEuskara\\n\\nفارسی\\n\\nFrançais\\n\\nGalego\\n\\nGĩkũyũ\\n\\n한국어\\n\\nՀայերեն\\n\\nहिन्दी\\n\\nHrvatski\\n\\nIdo\\n\\nBahasa Indonesia\\n\\nÍslenska\\n\\nItaliano\\n\\nעברית\\n\\nJawa\\n\\nಕನ್ನಡ\\n\\nҚазақша\\n\\nLatina\\n\\nLatviešu\\n\\nLëtzebuergesch\\n\\nLimburgs\\n\\nLingua Franca Nova\\n\\nMagyar\\n\\nМакедонски\\n\\nമലയാളം\\n\\nმარგალური\\n\\nمصرى\\n\\nBahasa Melayu\\n\\nМонгол\\n\\nNederlands\\n\\n日本語\\n\\nNorsk bokmål\\n\\nNorsk nynorsk\\n\\nNovial\\n\\nOccitan\\n\\nپښتو\\n\\nPolski\\n\\nPortuguês\\n\\nRomână\\n\\nРусский\\n\\nShqip\\n\\nSimple English\\n\\nSlovenčina\\n\\nSlovenščina\\n\\nکوردی\\n\\nСрпски / srpski\\n\\nSrpskohrvatski / српскохрватски\\n\\nSuomi\\n\\nSvenska\\n\\nTagalog\\n\\nதமிழ்\\n\\nไทย\\n\\nTürkçe\\n\\nУкраїнська\\n\\nاردو\\n\\nTiếng Việt\\n\\nWinaray\\n\\n吴语\\n\\nייִדיש\\n\\n粵語\\n\\n中文\\n\\nEdit links\\n\\nArticle\\n\\nTalk\\n\\nEnglish\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nTools\\n\\nTools\\n\\nActions\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nGeneral\\n\\nWhat links here\\n\\nRelated changes\\n\\nUpload file\\n\\nSpecial pages\\n\\nPermanent link\\n\\nPage information\\n\\nCite this page\\n\\nWikidata item\\n\\nPrint/export\\n\\nDownload as PDF\\n\\nPrintable version\\n\\nIn other projects\\n\\nWikimedia Commons\\n\\nWikibooks\\n\\nWikiversity\\n\\nFrom Wikipedia, the free encyclopedia\\n\\nComputational analysis of large, complex sets of biological data\\n\\nFor the journal, see \\n\\nBioinformatics (journal).\\n\\n§\\xa0Sequence analysis for further information.\\n\\nNational Center for Biotechnology Information website)\\n\\nNot to be confused with \\n\\nBiological computation or\\n\\nGenetic algorithm.\\n\\n/ˌbaɪ.oʊˌɪnfərˈmætɪks/ (listen)) is an\\n\\ninterdisciplinary field of\\n\\nscience that develops methods and\\n\\nsoftware tools for understanding\\n\\nbiological data, especially when the data sets are large and complex. Bioinformatics uses\\n\\nbiology,\\n\\nchemistry,\\n\\nphysics,\\n\\ncomputer science,\\n\\ncomputer programming,\\n\\ninformation engineering,\\n\\nmathematics and\\n\\nstatistics to analyze and interpret\\n\\nbiological data. The subsequent process of analyzing and interpreting data is referred to as\\n\\ncomputational biology.\\n\\ncomputer simulation analyses of biological queries. They include reused specific analysis \"pipelines\", particularly in the field of\\n\\ngenomics, such as by the identification of\\n\\ngenes and single\\n\\nnucleotide polymorphisms (\\n\\nSNPs). These pipelines are used to better understand the genetic basis of disease, unique adaptations, desirable properties (esp. in agricultural species), or differences between populations. Bioinformatics also includes\\n\\nproteomics, which tries to understand the organizational principles within\\n\\nnucleic acid and\\n\\nprotein sequences.\\n\\n[1]\\n\\nsignal processing allow extraction of useful results from large amounts of raw data. In the field of genetics, it aids in sequencing and annotating genomes and their observed\\n\\nmutations. Bioinformatics includes\\n\\ntext mining of biological literature and the development of biological and gene\\n\\nontologies to organize and query biological data. It also plays a role in the analysis of gene and protein expression and regulation. Bioinformatics tools aid in comparing, analyzing and interpreting genetic and genomic data and more generally in the understanding of evolutionary aspects of molecular biology. At a more integrative level, it helps analyze and catalogue the biological pathways and networks that are an important part of\\n\\nsystems biology. In\\n\\nstructural biology, it aids in the simulation and modeling of DNA,\\n\\n[2]\\n\\n[2]\\n\\n[3]\\n\\n[4]\\n\\n[5]\\n\\n[6]\\n\\n[7]\\n\\n[8]\\n\\nHistory[edit]\\n\\nPaulien Hogeweg and\\n\\nBen Hesper in 1970, to refer to the study of information processes in biotic systems.\\n\\n[9]\\n\\n[10]\\n\\n[11]\\n\\n[12]\\n\\n[13]\\n\\nbiochemistry (the study of chemical processes in biological systems).\\n\\n[10]\\n\\nBioinformatics and computational biology involved the analysis of biological data, particularly DNA, RNA, and protein sequences. The field of bioinformatics experienced explosive growth starting in the mid-1990s, driven largely by the Human Genome Project and by rapid advances in DNA sequencing technology.\\n\\nalgorithms from\\n\\ngraph theory,\\n\\nartificial intelligence,\\n\\nsoft computing,\\n\\ndata mining,\\n\\nimage processing, and\\n\\ncomputer simulation. The algorithms in turn depend on theoretical foundations such as\\n\\ndiscrete mathematics,\\n\\ncontrol theory,\\n\\nsystem theory,\\n\\ninformation theory, and\\n\\nstatistics.\\n\\nSequences[edit]\\n\\nThere has been a tremendous advance in speed and cost reduction since the completion of the Human Genome Project, with some labs able to sequence over 100,000 billion bases each year, and a full genome can be sequenced for $1,000 or less.[14]\\n\\nprotein sequences became available after\\n\\nFrederick Sanger determined the sequence of\\n\\ninsulin in the early 1950s. Comparing multiple sequences manually turned out to be impractical.\\n\\nMargaret Oakley Dayhoff, a pioneer in the field,\\n\\n[15]\\n\\n[16]\\n\\nmolecular evolution.\\n\\n[17]\\n\\nElvin A. Kabat, who pioneered biological sequence analysis in 1970 with his comprehensive volumes of antibody sequences released online with Tai Te Wu between 1980 and 1991.\\n\\n[18]\\n\\nIn the 1970s, new techniques for sequencing DNA were applied to bacteriophage MS2 and øX174, and the extended nucleotide sequences were then parsed with informational and statistical algorithms.  These studies illustrated that well known features, such as the coding segments and the triplet code, are revealed in straightforward statistical analyses and were the proof of the concept that bioinformatics would be insightful.[19][20]\\n\\nGoals[edit]\\n\\nIn order to study how normal cellular activities are altered in different disease states, raw biological data must be combined to form a comprehensive picture of these activities. Therefore[when?], the field of bioinformatics has evolved such that the most pressing task now involves the analysis and interpretation of various types of data. This also includes nucleotide and amino acid sequences, protein domains, and protein structures.[21]\\n\\nImportant sub-disciplines within bioinformatics and computational biology include:\\n\\nDevelopment and implementation of computer programs to efficiently access, manage, and use various types of information.\\n\\nDevelopment of new mathematical algorithms and statistical measures to assess relationships among members of large data sets. For example, there are methods to locate a gene within a sequence, to predict protein structure and/or function, and to cluster protein sequences into families of related sequences.\\n\\npattern recognition,\\n\\ndata mining,\\n\\nmachine learning algorithms, and\\n\\nvisualization. Major research efforts in the field include\\n\\nsequence alignment,\\n\\ngene finding,\\n\\ngenome assembly,\\n\\ndrug design,\\n\\ndrug discovery,\\n\\nprotein structure alignment,\\n\\nprotein structure prediction, prediction of\\n\\ngene expression and\\n\\nprotein–protein interactions,\\n\\ngenome-wide association studies, the modeling of\\n\\nevolution and\\n\\ncell division/mitosis.\\n\\nBioinformatics entails the creation and advancement of databases, algorithms, computational and statistical techniques, and theory to solve formal and practical problems arising from the management and analysis of biological data.\\n\\nOver the past few decades, rapid developments in genomic and other molecular research technologies and developments in information technologies have combined to produce a tremendous amount of information related to molecular biology. Bioinformatics is the name given to these mathematical and computing approaches used to glean understanding of biological processes.\\n\\nCommon activities in bioinformatics include mapping and analyzing DNA and protein sequences, aligning DNA and protein sequences to compare them, and creating and viewing 3-D models of protein structures.\\n\\nSequence analysis[edit]\\n\\nMain articles: \\n\\nSequence alignment,\\n\\nSequence database, and\\n\\nAlignment-free sequence analysis\\n\\nPhage Φ-X174 was\\n\\nsequenced in 1977,\\n\\n[22]\\n\\nDNA sequences of thousands of organisms have been decoded and stored in databases. This sequence information is analyzed to determine genes that encode\\n\\nproteins, RNA genes, regulatory sequences, structural motifs, and repetitive sequences. A comparison of genes within a\\n\\nspecies or between different species can show similarities between protein functions, or relations between species (the use of\\n\\nmolecular systematics to construct\\n\\nphylogenetic trees). With the growing amount of data, it long ago became impractical to analyze DNA sequences manually.\\n\\nComputer programs such as\\n\\nBLAST are used routinely to search sequences—as of 2008, from more than 260,000 organisms, containing over 190 billion\\n\\nnucleotides.\\n\\n[23]\\n\\nDNA sequencing[edit]\\n\\nMain article: \\n\\nDNA sequencing\\n\\nBefore sequences can be analyzed, they are obtained from a data storage bank, such as GenBank. DNA sequencing is still a non-trivial problem as the raw data may be noisy or affected by weak signals. Algorithms have been developed for base calling for the various experimental approaches to DNA sequencing.\\n\\nSequence assembly[edit]\\n\\nMain article: \\n\\nSequence assembly\\n\\nMost DNA sequencing techniques produce short fragments of sequence that need to be assembled to obtain complete gene or genome sequences. The shotgun sequencing technique (used by The Institute for Genomic Research (TIGR) to sequence the first bacterial genome, Haemophilus influenzae)[24] generates the sequences of many thousands of small DNA fragments (ranging from 35 to 900 nucleotides long, depending on the sequencing technology). The ends of these fragments overlap and, when aligned properly by a genome assembly program, can be used to reconstruct the complete genome. Shotgun sequencing yields sequence data quickly, but the task of assembling the fragments can be quite complicated for larger genomes. For a genome as large as the human genome, it may take many days of CPU time on large-memory, multiprocessor computers to assemble the fragments, and the resulting assembly usually contains numerous gaps that must be filled in later. Shotgun sequencing is the method of choice for virtually all genomes sequenced (rather than chain-termination or chemical degradation methods), and genome assembly algorithms are a critical area of bioinformatics research.\\n\\nSee also: \\n\\nsequence analysis,\\n\\nsequence mining,\\n\\nsequence profiling tool, and\\n\\nsequence motif\\n\\nGenome annotation[edit]\\n\\nMain article: \\n\\nGene prediction\\n\\nIn genomics, annotation refers to the process of marking the stop and start regions of genes and other biological features in a sequenced DNA sequence. Many genomes are too large to be annotated by hand. As the rate of sequencing exceeds the rate of genome annotation, genome annotation has become the new bottleneck in bioinformatics[when?].\\n\\nGenome annotation can be classified into three levels: the nucleotide, protein, and process levels.\\n\\nGene finding is a chief aspect of nucleotide-level annotation. For complex genomes, a combination of ab initio gene prediction and sequence comparison with expressed sequence databases and other organisms can be successful. Nucleotide-level annotation also allows the integration of genome sequence with other genetic and physical maps of the genome.\\n\\nThe principal aim of protein-level annotation is to assign function to the protein products of the genome. Databases of protein sequences and functional domains and motifs are used for this type of annotation. About half of the predicted proteins in a new genome sequence tend to have no obvious function.\\n\\nUnderstanding the function of genes and their products in the context of cellular and organismal physiology is the goal of process-level annotation. An obstacle of process-level annotation has been the inconsistency of terms used by different model systems. The Gene Ontology Consortium is helping to solve this problem.[25]\\n\\n[24]\\n\\nThe Institute for Genomic Research, which performed the first complete sequencing and analysis of the genome of a\\n\\nfree-living organism, the bacterium\\n\\nHaemophilus influenzae\\n\\n[24]\\n\\nGeneMark program trained to find protein-coding genes in\\n\\nHaemophilus influenzae\\n\\nFollowing the goals that the Human Genome Project left to achieve after its closure in 2003, the ENCODE project was developed by the National Human Genome Research Institute. This project is a collaborative data collection of the functional elements of the human genome that uses next-generation DNA-sequencing technologies and genomic tiling arrays, technologies able to automatically generate large amounts of data at a dramatically reduced per-base cost but with the same accuracy (base call error) and fidelity (assembly error).\\n\\nGene function prediction[edit]\\n\\nhomology), other properties of sequences can be used to predict the function of genes. In fact, most\\n\\namino acids predicts\\n\\ntransmembrane segments in proteins. However, protein function prediction can also use external information such as gene (or protein)\\n\\nexpression data,\\n\\nprotein structure, or\\n\\nprotein-protein interactions.\\n\\n[26]\\n\\nComputational evolutionary biology[edit]\\n\\nFurther information: \\n\\nComputational phylogenetics\\n\\nEvolutionary biology is the study of the origin and descent of species, as well as their change over time. Informatics has assisted evolutionary biologists by enabling researchers to:\\n\\ntrace the evolution of a large number of organisms by measuring changes in their DNA, rather than through physical taxonomy or physiological observations alone,\\n\\ncompare entire genomes, which permits the study of more complex evolutionary events, such as gene duplication, horizontal gene transfer, and the prediction of factors important in bacterial speciation,\\n\\nbuild complex computational population genetics models to predict the outcome of the system over time[27]\\n\\ntrack and share information on an increasingly large number of species and organisms\\n\\nFuture work endeavours to reconstruct the now more complex tree of life.[according to whom?]\\n\\nComparative genomics[edit]\\n\\nMain article: \\n\\nComparative genomics\\n\\ngenes (\\n\\northology analysis) or other genomic features in different organisms. Intergenomic maps are made to trace the evolutionary processes responsible for the divergence of two genomes. A multitude of evolutionary events acting at various organizational levels shape genome evolution. At the lowest level, point mutations affect individual nucleotides. At a higher level, large chromosomal segments undergo duplication, lateral transfer, inversion, transposition, deletion and insertion.\\n\\n[28]\\n\\nendosymbiosis that lead to rapid speciation. The complexity of genome evolution poses many exciting challenges to developers of mathematical models and algorithms, who have recourse to a spectrum of algorithmic, statistical and mathematical techniques, ranging from exact,\\n\\nheuristics, fixed parameter and\\n\\napproximation algorithms for problems based on parsimony models to\\n\\nMarkov chain Monte Carlo algorithms for\\n\\nBayesian analysis of problems based on probabilistic models.\\n\\nMany of these studies are based on the detection of sequence homology to assign sequences to protein families.[29]\\n\\nPan genomics[edit]\\n\\nMain article: \\n\\nPan-genome\\n\\nPan genomics is a concept introduced in 2005 by Tettelin and Medini. Pan genome is the complete gene repertoire of a particular monophyletic taxonomic group. Although initially applied to closely related strains of a species, it can be applied to a larger context like genus, phylum, etc. It is divided in two parts: the Core genome, a set of genes common to all the genomes under study (often housekeeping genes vital for survival), and the Dispensable/Flexible genome: a set of genes not present in all but one or some genomes under study. A bioinformatics tool BPGA can be used to characterize the Pan Genome of bacterial species.[30]\\n\\nGenetics of disease[edit]\\n\\nMain article: \\n\\nGenome-wide association studies\\n\\nMendelian inheritance has been observed for over 3,000 disorders that have been identified at the\\n\\nOnline Mendelian Inheritance in Man database, but complex diseases are more difficult. Association studies have found many individual genetic regions that individually are weakly associated with complex diseases (such as\\n\\ninfertility,\\n\\n[31]\\n\\nbreast cancer\\n\\n[32]\\n\\nAlzheimer\\'s disease\\n\\n[33]\\n\\n[34]\\n\\n[35]\\n\\n[36]\\n\\n[37]\\n\\nRare variants may account for some of the\\n\\nmissing heritability.\\n\\n[38]\\n\\nwhole genome sequencing studies have rapidly sequenced millions of whole genomes, and such studies have identified hundreds of millions of\\n\\nrare variants.\\n\\n[39]\\n\\nFunctional annotations predict the effect or function of a genetic variant and help to prioritize rare functional variants, and incorporating these annotations can effectively boost the power of genetic association of rare variants analysis of whole genome sequencing studies.\\n\\n[40]\\n\\n[41]\\n\\n[42]\\n\\n[43]\\n\\nAnalysis of mutations in cancer[edit]\\n\\nMain article: \\n\\nOncogenomics\\n\\ncancer, the genomes of affected cells are rearranged in complex or unpredictable ways. In addition to\\n\\nsingle-nucleotide polymorphism arrays identifying\\n\\npoint mutations that cause cancer,\\n\\noligonucleotide microarrays can be used to identify chromosomal gains and losses (called\\n\\ncomparative genomic hybridization). These detection methods generate\\n\\nterabytes of data per experiment. The data is often found to contain considerable variability, or\\n\\nnoise, and thus\\n\\nHidden Markov model and change-point analysis methods are being developed to infer real\\n\\ncopy number changes.\\n\\ncitation needed\\n\\nTwo important principles can be used to identify cancer by mutations in the exome. First, cancer is a disease of accumulated somatic mutations in genes. Second, cancer contains driver mutations which need to be distinguished from passengers.[44]\\n\\nFurther improvements in bioinformatics could allow for classifying types of cancer by analysis of cancer driven mutations in the genome. Furthermore, tracking of patients while the disease progresses may be possible in the future with the sequence of cancer samples. Another type of data that requires novel informatics development is the analysis of lesions found to be recurrent among many tumors.[45]\\n\\nGene and protein expression[edit]\\n\\nAnalysis of gene expression[edit]\\n\\nexpression of many genes can be determined by measuring\\n\\nmRNA levels with multiple techniques including\\n\\nmicroarrays,\\n\\nexpressed cDNA sequence tag (EST) sequencing,\\n\\nserial analysis of gene expression (SAGE) tag sequencing,\\n\\nmassively parallel signature sequencing (MPSS),\\n\\nRNA-Seq, also known as \"Whole Transcriptome Shotgun Sequencing\" (WTSS), or various applications of multiplexed in-situ hybridization. All of these techniques are extremely noise-prone and/or subject to bias in the biological measurement, and a major research area in computational biology involves developing statistical tools to separate\\n\\nsignal from\\n\\nnoise in high-throughput gene expression studies.\\n\\n[46]\\n\\nepithelial cells to data from non-cancerous cells to determine the transcripts that are up-regulated and down-regulated in a particular population of cancer cells.\\n\\nAnalysis of protein expression[edit]\\n\\nProtein microarrays and high throughput (HT)\\n\\nmass spectrometry (MS) can provide a snapshot of the proteins present in a biological sample. The former approach faces similar problems as with microarrays targeted at mRNA, the latter involves the problem of matching large amounts of mass data against predicted masses from protein sequence databases, and the complicated statistical analysis of samples when multiple incomplete peptides from each protein are detected. Cellular protein localization in a tissue context can be achieved through affinity\\n\\nproteomics displayed as spatial data based on\\n\\nimmunohistochemistry and\\n\\ntissue microarrays.\\n\\n[47]\\n\\nAnalysis of regulation[edit]\\n\\nGene regulation is a complex process where a signal, such as an extracellular signal such as a hormone, eventually leads to an increase or decrease in the activity of one or more proteins. Bioinformatics techniques have been applied to explore various steps in this process.\\n\\nFor example, gene expression can be regulated by nearby elements in the genome. Promoter analysis involves the identification and study of sequence motifs in the DNA surrounding the protein-coding region of a gene. These motifs influence the extent to which that region is transcribed into mRNA. Enhancer elements far away from the promoter can also regulate gene expression, through three-dimensional looping interactions. These interactions can be determined by bioinformatic analysis of chromosome conformation capture experiments.\\n\\nmicroarray data from a wide variety of states of an organism to form hypotheses about the genes involved in each state. In a single-cell organism, one might compare stages of the\\n\\ncell cycle, along with various stress conditions (heat shock, starvation, etc.).\\n\\nClustering algorithms can be then applied to expression data to determine which genes are co-expressed. For example, the upstream regions (promoters) of co-expressed genes can be searched for over-represented\\n\\nregulatory elements. Examples of clustering algorithms applied in gene clustering are\\n\\nk-means clustering,\\n\\nself-organizing maps (SOMs),\\n\\nhierarchical clustering, and\\n\\nconsensus clustering methods.\\n\\nAnalysis of cellular organization[edit]\\n\\nSeveral approaches have been developed to analyze the location of organelles, genes, proteins, and other components within cells. A gene ontology category, cellular component, has been devised to capture subcellular localization in many biological databases.\\n\\nMicroscopy and image analysis[edit]\\n\\nMicroscopic pictures allow for the location of organelles as well as molecules, which may be the source of abnormalities in diseases.\\n\\nProtein localization[edit]\\n\\nprotein function prediction. For instance, if a protein is found in the\\n\\nnucleus it may be involved in\\n\\ngene regulation or\\n\\nsplicing. By contrast, if a protein is found in\\n\\nmitochondria, it may be involved in\\n\\nrespiration or other\\n\\nmetabolic processes. There are well developed\\n\\nprotein subcellular localization prediction resources available, including protein subcellular location databases, and prediction tools.\\n\\n[48]\\n\\n[49]\\n\\nNuclear organization of chromatin[edit]\\n\\nMain article: \\n\\nNuclear organization\\n\\nchromosome conformation capture experiments, such as\\n\\nHi-C (experiment) and\\n\\nChIA-PET, can provide information on the three-dimensional structure and\\n\\nnuclear organization of\\n\\nchromatin. Bioinformatic challenges in this field include partitioning the genome into domains, such as\\n\\nTopologically Associating Domains (TADs), that are organised together in three-dimensional space.\\n\\n[50]\\n\\nStructural bioinformatics[edit]\\n\\nMain articles: \\n\\nStructural bioinformatics and\\n\\nProtein structure prediction\\n\\nSee also: \\n\\nStructural motif and\\n\\nStructural domain\\n\\nFinding the structure of proteins important application of bioinformatics. The Critical Assessment of Protein Structure Prediction (CASP) is an open competition where worldwide research groups submit protein models for evaluating unknown protein models.[51][52]\\n\\nAmino acid sequence[edit]\\n\\namino acid sequence of a protein is called the\\n\\nprimary structure, can be easily determined from the sequence of\\n\\ncodons on the DNA gene that codes for it. In most proteins, the primary structure uniquely determines the 3-dimensional structure of a protein in its native environment. An exception is the\\n\\nmisfolded protein involved in\\n\\nbovine spongiform encephalopathy. This structure is linked to the function of the protein. Additional structural information includes the\\n\\nsecondary\\n\\ntertiary\\n\\nquaternary\\n\\ncitation needed\\n\\nHomology[edit]\\n\\nIn the genomic branch of bioinformatics, homology is used to predict the function of a gene: if the sequence of gene A, whose function is known, is homologous to the sequence of gene B, whose function is unknown, one could infer that B may share A\\'s function. In structural bioinformatics, homology is used to determine which parts of a protein are important in structure formation and interaction with other proteins. Homology modeling is used to predict the structure of an unknown protein from existing homologous proteins.\\n\\nOne example of this is hemoglobin in humans and the hemoglobin in legumes (leghemoglobin), which are distant relatives from the same protein superfamily. Both serve the same purpose of transporting oxygen in the organism. Although both of these proteins have completely different amino acid sequences, their protein structures are virtually identical, which reflects their near identical purposes and shared ancestor.[53]\\n\\nOther techniques for predicting protein structure include protein threading and de novo (from scratch) physics-based modeling.\\n\\nAnother aspect of structural bioinformatics include the use of protein structures for Virtual Screening models such as Quantitative Structure-Activity Relationship models and proteochemometric models (PCM). Furthermore, a protein\\'s crystal structure can be used in simulation of for example ligand-binding studies and in silico mutagenesis studies.\\n\\ndeep-learning algorithms-based software called\\n\\nAlphaFold, developed by Google\\'s\\n\\nDeepMind, greatly outperforms all other prediction software methods\\n\\n[54]\\n\\nhow?\\n\\n[55]\\n\\nNetwork and systems biology[edit]\\n\\nMain articles: \\n\\nComputational systems biology,\\n\\nBiological network, and\\n\\nInteractome\\n\\nNetwork analysis seeks to understand the relationships within biological networks such as metabolic or protein–protein interaction networks. Although biological networks can be constructed from a single type of molecule or entity (such as genes), network biology often attempts to integrate many different data types, such as proteins, small molecules, gene expression data, and others, which are all connected physically, functionally, or both.\\n\\ncomputer simulations of\\n\\ncellular subsystems (such as the\\n\\nnetworks of metabolites and\\n\\nenzymes that comprise\\n\\nmetabolism,\\n\\nsignal transduction pathways and\\n\\ngene regulatory networks) to both analyze and visualize the complex connections of these cellular processes.\\n\\nArtificial life or virtual evolution attempts to understand evolutionary processes via the computer simulation of simple (artificial) life forms.\\n\\nMolecular interaction networks[edit]\\n\\nTreponema pallidum\\n\\nsyphilis and other diseases.\\n\\n[56]\\n\\nMain articles: \\n\\nProtein–protein interaction prediction and\\n\\ninteractome\\n\\nTens of thousands of three-dimensional protein structures have been determined by X-ray crystallography and protein nuclear magnetic resonance spectroscopy (protein NMR) and a central question in structural bioinformatics is whether it is practical to predict possible protein–protein interactions only based on these 3D shapes, without performing protein–protein interaction experiments. A variety of methods have been developed to tackle the protein–protein docking problem, though it seems that there is still much work to be done in this field.\\n\\nOther interactions encountered in the field include Protein–ligand (including drug) and protein–peptide. Molecular dynamic simulation of movement of atoms about rotatable bonds is the fundamental principle behind computational algorithms, termed docking algorithms, for studying molecular interactions.\\n\\nOthers[edit]\\n\\nLiterature analysis[edit]\\n\\nMain articles: \\n\\nText mining and\\n\\nBiomedical text mining\\n\\nThe enormous number of published literature makes it virtually impossible for individuals to read every paper, resulting in disjointed sub-fields of research. Literature analysis aims to employ computational and statistical linguistics to mine this growing library of text resources. For example:\\n\\nAbbreviation recognition – identify the long-form and abbreviation of biological terms\\n\\nNamed-entity recognition – recognizing biological terms such as gene names\\n\\nProtein–protein interaction – identify which proteins interact with which proteins from text\\n\\nThe area of research draws from statistics and computational linguistics.\\n\\nHigh-throughput image analysis[edit]\\n\\nComputational technologies are used to automate the processing, quantification and analysis of large amounts of high-information-content biomedical imagery. Modern image analysis systems can improve an observer\\'s accuracy, objectivity, or speed. Image analysis is important for both diagnostics and research. Some examples are:\\n\\nhigh-throughput and high-fidelity quantification and sub-cellular localization (high-content screening, cytohistopathology, Bioimage informatics)\\n\\nmorphometrics\\n\\nclinical image analysis and visualization\\n\\ndetermining the real-time air-flow patterns in breathing lungs of living animals\\n\\nquantifying occlusion size in real-time imagery from the development of and recovery during arterial injury\\n\\nmaking behavioral observations from extended video recordings of laboratory animals\\n\\ninfrared measurements for metabolic activity determination\\n\\ninferring clone overlaps in DNA mapping, e.g. the Sulston score\\n\\nHigh-throughput single cell data analysis[edit]\\n\\nMain article: \\n\\nFlow cytometry bioinformatics\\n\\nComputational techniques are used to analyse high-throughput, low-measurement single cell data, such as that obtained from flow cytometry. These methods typically involve finding populations of cells that are relevant to a particular disease state or experimental condition.\\n\\nBiodiversity informatics[edit]\\n\\nMain article: \\n\\nBiodiversity informatics\\n\\nbiodiversity data, such as\\n\\ntaxonomic databases, or\\n\\nmicrobiome data. Examples of such analyses include\\n\\nphylogenetics,\\n\\nniche modelling,\\n\\nspecies richness mapping,\\n\\nDNA barcoding, or\\n\\nspecies identification tools.\\n\\nOntologies and data integration[edit]\\n\\nBiological ontologies are directed acyclic graphs of controlled vocabularies. They create categories for biological concepts and descriptions so they can be easily analyzed with computers. When categorised in this way, it is possible to gain added value from holistic and integrated analysis.[citation needed]\\n\\nThe OBO Foundry was an effort to standardise certain ontologies. One of the most widespread is the Gene ontology which describes gene function. There are also ontologies which describe phenotypes.\\n\\nDatabases[edit]\\n\\nMain articles: \\n\\nList of biological databases and\\n\\nBiological database\\n\\nDatabases are essential for bioinformatics research and applications. Databases exist for many different information types, including DNA and protein sequences, molecular structures, phenotypes and biodiversity. Databases can contain both empirical data (obtained directly from experiments) and predicted data (obtained from analysis of existing data). They may be specific to a particular organism, pathway or molecule of interest. Alternatively, they can incorporate data compiled from multiple other databases. Databases can have different formats, access mechanisms, and be public or private.\\n\\nSome of the most commonly used databases are listed below:\\n\\nUsed in biological sequence analysis: Genbank, UniProt\\n\\nUsed in structure analysis: Protein Data Bank (PDB)\\n\\nUsed in finding Protein Families and Motif Finding: InterPro, Pfam\\n\\nUsed for Next Generation Sequencing: Sequence Read Archive\\n\\nUsed in Network Analysis: Metabolic Pathway Databases (KEGG, BioCyc), Interaction Analysis Databases, Functional Networks\\n\\nUsed in design of synthetic genetic circuits: GenoCAD[citation needed]\\n\\nSoftware and tools[edit]\\n\\nSoftware tools for bioinformatics include simple command-line tools, more complex graphical programs, and standalone web-services. They are made by bioinformatics companies or by public institutions.\\n\\nOpen-source bioinformatics software[edit]\\n\\nMain article: \\n\\nList of bioinformatics software\\n\\nfree and open-source software tools have existed and continued to grow since the 1980s.\\n\\n[57]\\n\\nalgorithms for the analysis of emerging types of biological readouts, the potential for innovative\\n\\nin silico\\n\\nopen code bases have created opportunities for research groups to contribute to both bioinformatics regardless of\\n\\nfunding. The open source tools often act as incubators of ideas, or community-supported\\n\\nplug-ins in commercial applications. They may also provide\\n\\nde facto\\n\\nBioconductor,\\n\\nBioPerl,\\n\\nBiopython,\\n\\nBioJava,\\n\\nBioJS,\\n\\nBioRuby,\\n\\nBioclipse,\\n\\nEMBOSS,\\n\\n.NET Bio,\\n\\nOrange with its bioinformatics add-on,\\n\\nApache Taverna,\\n\\nUGENE and\\n\\nGenoCAD.\\n\\nThe non-profit Open Bioinformatics Foundation[57] and the annual Bioinformatics Open Source Conference promote open-source bioinformatics software.[58]\\n\\nWeb services in bioinformatics[edit]\\n\\nSOAP- and REST-based interfaces have been developed to allow client computers to use algorithms, data and computing resources from servers in other parts of the world. The main advantage are that end users do not have to deal with software and database maintenance overheads.\\n\\nEBI into three categories:\\n\\nSSS (Sequence Search Services),\\n\\nMSA (Multiple Sequence Alignment), and\\n\\nBSA (Biological Sequence Analysis).\\n\\n[59]\\n\\nservice-oriented bioinformatics resources demonstrate the applicability of web-based bioinformatics solutions, and range from a collection of standalone tools with a common data format under a single web-based interface, to integrative, distributed and extensible\\n\\nbioinformatics workflow management systems.\\n\\nBioinformatics workflow management systems[edit]\\n\\nMain article: \\n\\nBioinformatics workflow management systems\\n\\nA bioinformatics workflow management system is a specialized form of a workflow management system designed specifically to compose and execute a series of computational or data manipulation steps, or a workflow, in a Bioinformatics application. Such systems are designed to\\n\\nprovide an easy-to-use environment for individual application scientists themselves to create their own workflows,\\n\\nprovide interactive tools for the scientists enabling them to execute their workflows and view their results in real-time,\\n\\nsimplify the process of sharing and reusing workflows between the scientists, and\\n\\nenable scientists to track the provenance of the workflow execution results and the workflow creation steps.\\n\\nGalaxy,\\n\\nKepler,\\n\\nTaverna,\\n\\nUGENE,\\n\\nAnduril,\\n\\nHIVE.\\n\\nBioCompute and BioCompute Objects[edit]\\n\\nUS Food and Drug Administration sponsored a conference held at the\\n\\nNational Institutes of Health Bethesda Campus to discuss reproducibility in bioinformatics.\\n\\n[60]\\n\\n[61]\\n\\nHuman Variome Project and the\\n\\nEuropean Federation for Medical Informatics, and research institutions including\\n\\nStanford, the\\n\\nNew York Genome Center, and the\\n\\nGeorge Washington University.\\n\\nIt was decided that the BioCompute paradigm would be in the form of digital \\'lab notebooks\\' which allow for the reproducibility, replication, review, and reuse, of bioinformatics protocols. This was proposed to enable greater continuity within a research group over the course of normal personnel flux while furthering the exchange of ideas between groups. The US FDA funded this work so that information on pipelines would be more transparent and accessible to their regulatory staff.[62]\\n\\nIn 2016, the group reconvened at the NIH in Bethesda and discussed the potential for a BioCompute Object, an instance of the BioCompute paradigm. This work was copied as both a \"standard trial use\" document and a preprint paper uploaded to bioRxiv. The BioCompute object allows for the JSON-ized record to be shared among employees, collaborators, and regulators.[63][64]\\n\\nEducation platforms[edit]\\n\\nmasters degree at many universities. The computational nature of bioinformatics lends it to\\n\\ncomputer-aided and online learning.\\n\\n[65]\\n\\n[66]\\n\\nRosalind and online courses offered through the\\n\\nSwiss Institute of Bioinformatics Training Portal. The\\n\\nCanadian Bioinformatics Workshops provides videos and slides from training workshops on their website under a\\n\\nCreative Commons license. The 4273π project or 4273pi project\\n\\n[67]\\n\\nRaspberry Pi computers and has been used to teach adults and school pupils.\\n\\n[68]\\n\\n[69]\\n\\n[70]\\n\\n[71]\\n\\nMOOC platforms also provide online certifications in bioinformatics and related disciplines, including\\n\\nCoursera\\'s Bioinformatics Specialization (\\n\\nUC San Diego) and Genomic Data Science Specialization (\\n\\nJohns Hopkins) as well as\\n\\nEdX\\'s Data Analysis for Life Sciences XSeries (\\n\\nHarvard).\\n\\nConferences[edit]\\n\\nThere are several large conferences that are concerned with bioinformatics. Some of the most notable examples are Intelligent Systems for Molecular Biology (ISMB), European Conference on Computational Biology (ECCB), and Research in Computational Molecular Biology (RECOMB).\\n\\nSee also[edit]\\n\\nBiodiversity informatics\\n\\nBioinformatics companies\\n\\nComputational biology\\n\\nComputational biomodeling\\n\\nComputational genomics\\n\\nCyberbiosecurity\\n\\nFunctional genomics\\n\\nHealth informatics\\n\\nInternational Society for Computational Biology\\n\\nJumping library\\n\\nList of bioinformatics institutions\\n\\nList of open-source bioinformatics software\\n\\nList of bioinformatics journals\\n\\nMetabolomics\\n\\nNucleic acid sequence\\n\\nPhylogenetics\\n\\nProteomics\\n\\nGene Disease Database\\n\\nReferences[edit]\\n\\n^ .mw-parser-output cite.citation{font-style:inherit;word-wrap:break-word}.mw-parser-output .citation q{quotes:\"\\\\\"\"\"\\\\\"\"\"\\'\"\"\\'\"}.mw-parser-output .citation:target{background-color:rgba(0,127,255,0.133)}.mw-parser-output .id-lock-free a,.mw-parser-output .citation .cs1-lock-free a{background:url(\"//upload.wikimedia.org/wikipedia/commons/6/65/Lock-green.svg\")right 0.1em center/9px no-repeat}.mw-parser-output .id-lock-limited a,.mw-parser-output .id-lock-registration a,.mw-parser-output .citation .cs1-lock-limited a,.mw-parser-output .citation .cs1-lock-registration a{background:url(\"//upload.wikimedia.org/wikipedia/commons/d/d6/Lock-gray-alt-2.svg\")right 0.1em center/9px no-repeat}.mw-parser-output .id-lock-subscription a,.mw-parser-output .citation .cs1-lock-subscription a{background:url(\"//upload.wikimedia.org/wikipedia/commons/a/aa/Lock-red-alt-2.svg\")right 0.1em center/9px no-repeat}.mw-parser-output .cs1-ws-icon a{background:url(\"//upload.wikimedia.org/wikipedia/commons/4/4c/Wikisource-logo.svg\")right 0.1em center/12px no-repeat}.mw-parser-output .cs1-code{color:inherit;background:inherit;border:none;padding:inherit}.mw-parser-output .cs1-hidden-error{display:none;color:#d33}.mw-parser-output .cs1-visible-error{color:#d33}.mw-parser-output .cs1-maint{display:none;color:#3a3;margin-left:0.3em}.mw-parser-output .cs1-format{font-size:95%}.mw-parser-output .cs1-kern-left{padding-left:0.2em}.mw-parser-output .cs1-kern-right{padding-right:0.2em}.mw-parser-output .citation .mw-selflink{font-weight:inherit}Lesk AM (26 July 2013). \"Bioinformatics\". Encyclopaedia Britannica. Archived from the original on 14 April 2021. Retrieved 17 April 2017.\\n\\n^ a b Sim AY, Minary P, Levitt M (June 2012). \"Modeling nucleic acids\". Current Opinion in Structural Biology. 22 (3): 273–8. doi:10.1016/j.sbi.2012.03.012. PMC\\xa04028509. PMID\\xa022538125.\\n\\n^ Dawson WK, Maciejczyk M, Jankowska EJ, Bujnicki JM (July 2016). \"Coarse-grained modeling of RNA 3D structure\". Methods. 103: 138–56. doi:10.1016/j.ymeth.2016.04.026. PMID\\xa027125734.\\n\\n^ Kmiecik S, Gront D, Kolinski M, Wieteska L, Dawid AE, Kolinski A (July 2016). \"Coarse-Grained Protein Models and Their Applications\". Chemical Reviews. 116 (14): 7898–936. doi:10.1021/acs.chemrev.6b00163. PMID\\xa027333362.\\n\\n^ Wong KC (2016). Computational Biology and Bioinformatics: Gene Regulation. CRC Press/Taylor & Francis Group. ISBN\\xa09781498724975.\\n\\n^ Joyce AP, Zhang C, Bradley P, Havranek JJ (January 2015). \"Structure-based modeling of protein: DNA specificity\". Briefings in Functional Genomics. 14 (1): 39–49. doi:10.1093/bfgp/elu044. PMC\\xa04366589. PMID\\xa025414269.\\n\\n^ Spiga E, Degiacomi MT, Dal Peraro M (2014). \"New Strategies for Integrative Dynamic Modeling of Macromolecular Assembly\". In Karabencheva-Christova T (ed.). Biomolecular Modelling and Simulations. Advances in Protein Chemistry and Structural Biology. Vol.\\xa096. Academic Press. pp.\\xa077–111. doi:10.1016/bs.apcsb.2014.06.008. ISBN\\xa09780128000137. PMID\\xa025443955.\\n\\n^ Ciemny M, Kurcinski M, Kamel K, Kolinski A, Alam N, Schueler-Furman O, Kmiecik S (August 2018). \"Protein-peptide docking: opportunities and challenges\". Drug Discovery Today. 23 (8): 1530–1537. doi:10.1016/j.drudis.2018.05.006. PMID\\xa029733895.\\n\\n^ Ouzounis, C. A.; Valencia, A. (2003). \"Early bioinformatics: the birth of a discipline—a personal view\". Bioinformatics. 19 (17): 2176–2190. doi:10.1093/bioinformatics/btg309. PMID\\xa014630646.\\n\\n^ a b Hogeweg P (March 2011). Searls DB (ed.). \"The roots of bioinformatics in theoretical biology\". PLOS Computational Biology. 7 (3): e1002021. Bibcode:2011PLSCB...7E2021H. doi:10.1371/journal.pcbi.1002021. PMC\\xa03068925. PMID\\xa021483479.\\n\\n^ Hesper B, Hogeweg P (1970). \"Bio-informatica: een werkconcept\". Kameleon. 1 (6): 28–29.\\n\\n^ Hesper B, Hogeweg P (2021). \"Bio-informatics: a working concept. A translation of \"Bio-informatica: een werkconcept\" by B. Hesper and P. Hogeweg\". arXiv:2111.11832v1 [q-bio.OT].\\n\\n^ Hogeweg P (1978). \"Simulating the growth of cellular forms\". Simulation. 31 (3): 90–96. doi:10.1177/003754977803100305. S2CID\\xa061206099.\\n\\n^ Colby B (2022). \"Whole Genome Sequencing Cost\". Sequencing.com. Archived from the original on 15 March 2022. Retrieved 8 April 2022.\\n\\n^ Moody G (2004). Digital Code of Life: How Bioinformatics is Revolutionizing Science, Medicine, and Business. ISBN\\xa0978-0-471-32788-2.\\n\\n^ Dayhoff, M.O. (1966) Atlas of protein sequence and structure. National Biomedical Research Foundation, 215 pp.\\n\\n^ Eck RV, Dayhoff MO (April 1966). \"Evolution of the structure of ferredoxin based on living relics of primitive amino Acid sequences\". Science. 152 (3720): 363–6. Bibcode:1966Sci...152..363E. doi:10.1126/science.152.3720.363. PMID\\xa017775169. S2CID\\xa023208558.\\n\\n^ Johnson G, Wu TT (January 2000). \"Kabat database and its applications: 30 years after the first variability plot\". Nucleic Acids Research. 28 (1): 214–8. doi:10.1093/nar/28.1.214. PMC\\xa0102431. PMID\\xa010592229.\\n\\n^ Erickson JW, Altman GG (1979). \"A Search for Patterns in the Nucleotide Sequence of the MS2 Genome\". Journal of Mathematical Biology. 7 (3): 219–230. doi:10.1007/BF00275725. S2CID\\xa085199492.\\n\\n^ Shulman MJ, Steinberg CM, Westmoreland N (February 1981). \"The coding function of nucleotide sequences can be discerned by statistical analysis\". Journal of Theoretical Biology. 88 (3): 409–20. Bibcode:1981JThBi..88..409S. doi:10.1016/0022-5193(81)90274-5. PMID\\xa06456380.\\n\\n^ Xiong J (2006). Essential Bioinformatics. Cambridge, United Kingdom: Cambridge University Press. pp.\\xa04. ISBN\\xa0978-0-511-16815-4 – via Internet Archive.\\n\\n^ Sanger F, Air GM, Barrell BG, Brown NL, Coulson AR, Fiddes CA, et\\xa0al. (February 1977). \"Nucleotide sequence of bacteriophage phi X174 DNA\". Nature. 265 (5596): 687–95. Bibcode:1977Natur.265..687S. doi:10.1038/265687a0. PMID\\xa0870828. S2CID\\xa04206886.\\n\\n^ Benson DA, Karsch-Mizrachi I, Lipman DJ, Ostell J, Wheeler DL (January 2008). \"GenBank\". Nucleic Acids Research. 36 (Database issue): D25-30. doi:10.1093/nar/gkm929. PMC\\xa02238942. PMID\\xa018073190.\\n\\n^ a b c Fleischmann RD, Adams MD, White O, Clayton RA, Kirkness EF, Kerlavage AR, et\\xa0al. (July 1995). \"Whole-genome random sequencing and assembly of Haemophilus influenzae Rd\". Science. 269 (5223): 496–512. Bibcode:1995Sci...269..496F. doi:10.1126/science.7542800. PMID\\xa07542800.\\n\\n^ Stein, Lincoln (2001). \"Genome annotation: from sequence to biology\". Nature. 2 (7): 493–503. doi:10.1038/35080529. PMID\\xa011433356. S2CID\\xa012044602.\\n\\n^ Erdin S, Lisewski AM, Lichtarge O (April 2011). \"Protein function prediction: towards integration of similarity metrics\". Current Opinion in Structural Biology. 21 (2): 180–8. doi:10.1016/j.sbi.2011.02.001. PMC\\xa03120633. PMID\\xa021353529.\\n\\n^ Carvajal-Rodríguez A (March 2010). \"Simulation of genes and genomes forward in time\". Current Genomics. 11 (1): 58–61. doi:10.2174/138920210790218007. PMC\\xa02851118. PMID\\xa020808525.\\n\\n^ Brown TA (2002). \"Mutation, Repair and Recombination\". Genomes (2nd\\xa0ed.). Manchester (UK): Oxford.\\n\\n^ Carter NP, Fiegler H, Piper J (October 2002). \"Comparative analysis of comparative genomic hybridization microarray technologies: report of a workshop sponsored by the Wellcome Trust\". Cytometry. 49 (2): 43–8. doi:10.1002/cyto.10153. PMID\\xa012357458.\\n\\n^ Chaudhari NM, Gupta VK, Dutta C (April 2016). \"BPGA- an ultra-fast pan-genome analysis pipeline\". Scientific Reports. 6: 24373. Bibcode:2016NatSR...624373C. doi:10.1038/srep24373. PMC\\xa04829868. PMID\\xa027071527.\\n\\n^ Aston KI (May 2014). \"Genetic susceptibility to male infertility: news from genome-wide association studies\". Andrology. 2 (3): 315–21. doi:10.1111/j.2047-2927.2014.00188.x. PMID\\xa024574159. S2CID\\xa0206007180.\\n\\n^ Véron A, Blein S, Cox DG (2014). \"Genome-wide association studies and the clinic: a focus on breast cancer\". Biomarkers in Medicine. 8 (2): 287–96. doi:10.2217/bmm.13.121. PMID\\xa024521025.\\n\\n^ Tosto G, Reitz C (October 2013). \"Genome-wide association studies in Alzheimer\\'s disease: a review\". Current Neurology and Neuroscience Reports. 13 (10): 381. doi:10.1007/s11910-013-0381-0. PMC\\xa03809844. PMID\\xa023954969.\\n\\n^ Londin E, Yadav P, Surrey S, Kricka LJ, Fortina P (2013). \"Use of linkage analysis, genome-wide association studies, and next-generation sequencing in the identification of disease-causing mutations\". Pharmacogenomics. Methods in Molecular Biology. Vol.\\xa01015. pp.\\xa0127–46. doi:10.1007/978-1-62703-435-7_8. ISBN\\xa0978-1-62703-434-0. PMID\\xa023824853.\\n\\n^ Hindorff LA, Sethupathy P, Junkins HA, Ramos EM, Mehta JP, Collins FS, Manolio TA (June 2009). \"Potential etiologic and functional implications of genome-wide association loci for human diseases and traits\". Proceedings of the National Academy of Sciences of the United States of America. 106 (23): 9362–7. Bibcode:2009PNAS..106.9362H. doi:10.1073/pnas.0903103106. PMC\\xa02687147. PMID\\xa019474294.\\n\\n^ Hall LO (2010). \"Finding the right genes for disease and prognosis prediction\". 2010 International Conference on System Science and Engineering. pp.\\xa01–2. doi:10.1109/ICSSE.2010.5551766. ISBN\\xa0978-1-4244-6472-2. S2CID\\xa021622726.\\n\\n^ Manolio, Teri A.; Collins, Francis S.; Cox, Nancy J.; et\\xa0al. (October 2009). \"Finding the missing heritability of complex diseases\". Nature. 461 (7265): 747–753. Bibcode:2009Natur.461..747M. doi:10.1038/nature08494. PMC\\xa02831613. PMID\\xa019812666.\\n\\n^ Wainschtein, Pierrick; Jain, Deepti; Zheng, Zhili; et\\xa0al. (March 2022). \"Assessing the contribution of rare variants to complex trait heritability from whole-genome sequence data\". Nature Genetics. 54 (3): 263–273. doi:10.1038/s41588-021-00997-7. PMC\\xa09119698. PMID\\xa035256806.\\n\\n^ Taliun, Daniel; Harris, Daniel N.; Kessler, Michael D.; et\\xa0al. (February 2021). \"Sequencing of 53,831 diverse genomes from the NHLBI TOPMed Program\". Nature. 590 (7845): 290–299. Bibcode:2021Natur.590..290T. doi:10.1038/s41586-021-03205-y. PMC\\xa07875770. PMID\\xa033568819.\\n\\n^ Li, Xihao; Li, Zilin; Zhou, Hufeng; et\\xa0al. (September 2020). \"Dynamic incorporation of multiple in silico functional annotations empowers rare variant association analysis of large whole-genome sequencing studies at scale\". Nature Genetics. 52 (9): 969–983. doi:10.1038/s41588-020-0676-4. PMC\\xa07483769. PMID\\xa032839606.\\n\\n^ Li, Zilin; Li, Xihao; Zhou, Hufeng; et\\xa0al. (December 2022). \"A framework for detecting noncoding rare-variant associations of large-scale whole-genome sequencing studies\". Nature Methods. 19 (12): 1599–1611. doi:10.1038/s41592-022-01640-x. PMC\\xa010008172. PMID\\xa036303018. S2CID\\xa0243873361.\\n\\n^ \"STAARpipeline: an all-in-one rare-variant tool for biobank-scale whole-genome sequencing data\". Nature Methods. 19 (12): 1532–1533. December 2022. doi:10.1038/s41592-022-01641-w. PMID\\xa036316564. S2CID\\xa0253246835.\\n\\n^ Li, Xihao; Quick, Corbin; Zhou, Hufeng; Gaynor, Sheila M.; Liu, Yaowu; Chen, Han; Selvaraj, Margaret Sunitha; Sun, Ryan; Dey, Rounak; Arnett, Donna K.; Bielak, Lawrence F.; Bis, Joshua C.; Blangero, John; Boerwinkle, Eric; Bowden, Donald W.; Brody, Jennifer A.; Cade, Brian E.; Correa, Adolfo; Cupples, L. Adrienne; Curran, Joanne E.; de Vries, Paul S.; Duggirala, Ravindranath; Freedman, Barry I.; Göring, Harald H. H.; Guo, Xiuqing; Haessler, Jeffrey; Kalyani, Rita R.; Kooperberg, Charles; Kral, Brian G.; Lange, Leslie A.; Manichaikul, Ani; Martin, Lisa W.; McGarvey, Stephen T.; Mitchell, Braxton D.; Montasser, May E.; Morrison, Alanna C.; Naseri, Take; O’Connell, Jeffrey R.; Palmer, Nicholette D.; Peyser, Patricia A.; Psaty, Bruce M.; Raffield, Laura M.; Redline, Susan; Reiner, Alexander P.; Reupena, Muagututi’a Sefuiva; Rice, Kenneth M.; Rich, Stephen S.; Sitlani, Colleen M.; Smith, Jennifer A.; Taylor, Kent D.; Vasan, Ramachandran S.; Willer, Cristen J.; Wilson, James G.; Yanek, Lisa R.; Zhao, Wei; NHLBI Trans-Omics for Precision Medicine (TOPMed) Consortium; TOPMed Lipids Working Group; Rotter, Jerome I.; Natarajan, Pradeep; Peloso, Gina M.; Li, Zilin; Lin, Xihong (January 2023). \"Powerful, scalable and resource-efficient meta-analysis of rare variant associations in large whole genome sequencing studies\". Nature Genetics. 55 (1): 154–164. doi:10.1038/s41588-022-01225-6. PMC\\xa010084891. PMID\\xa036564505. S2CID\\xa0255084231.\\n\\n^ Vazquez M, de la Torre V, Valencia A (27 December 2012). \"Chapter 14: Cancer genome analysis\". PLOS Computational Biology. 8 (12): e1002824. Bibcode:2012PLSCB...8E2824V. doi:10.1371/journal.pcbi.1002824. PMC\\xa03531315. PMID\\xa023300415.\\n\\n^ Hye-Jung EC, Jaswinder K, Martin K, Samuel AA, Marco AM (2014). \"Second-Generation Sequencing for Cancer Genome Analysis\". In Dellaire G, Berman JN, Arceci RJ (eds.). Cancer Genomics. Boston (US): Academic Press. pp.\\xa013–30. doi:10.1016/B978-0-12-396967-5.00002-5. ISBN\\xa09780123969675.\\n\\n^ Grau J, Ben-Gal I, Posch S, Grosse I (July 2006). \"VOMBAT: prediction of transcription factor binding sites using variable order Bayesian trees\". Nucleic Acids Research. 34 (Web Server issue): W529-33. doi:10.1093/nar/gkl212. PMC\\xa01538886. PMID\\xa016845064.\\n\\n^ \"The Human Protein Atlas\". www.proteinatlas.org. Archived from the original on 4 March 2020. Retrieved 2 October 2017.\\n\\n^ \"The human cell\". www.proteinatlas.org. Archived from the original on 2 October 2017. Retrieved 2 October 2017.\\n\\n^ Thul PJ, Åkesson L, Wiking M, Mahdessian D, Geladaki A, Ait Blal H, et\\xa0al. (May 2017). \"A subcellular map of the human proteome\". Science. 356 (6340): eaal3321. doi:10.1126/science.aal3321. PMID\\xa028495876. S2CID\\xa010744558.\\n\\n^ Ay F, Noble WS (September 2015). \"Analysis methods for studying the 3D architecture of the genome\". Genome Biology. 16 (1): 183. doi:10.1186/s13059-015-0745-7. PMC\\xa04556012. PMID\\xa026328929.\\n\\n^ Kryshtafovych, A.; Schwede, T.; Topf, M.; Fidelis, K.; Moult, J. (2019). \"Critical Assessment of Methods of Protein Structure Prediction (CASP) – Round XIII\". Proteins. 87 (12): 1011–1020. doi:10.1002/prot.25823. PMC\\xa06927249. PMID\\xa031589781.\\n\\n^ \"Home - CASP14\". predictioncenter.org. Archived from the original on 30 January 2023. Retrieved 12 June 2023.\\n\\n^ Hoy JA, Robinson H, Trent JT, Kakar S, Smagghe BJ, Hargrove MS (August 2007). \"Plant hemoglobins: a molecular fossil record for the evolution of oxygen transport\". Journal of Molecular Biology. 371 (1): 168–79. doi:10.1016/j.jmb.2007.05.029. PMID\\xa017560601.\\n\\n^ Jumper, John; Evans, Richard; Pritzel, Alexander; Green, Tim; Figurnov, Michael; Ronneberger, Olaf; Tunyasuvunakool, Kathryn; Bates, Russ; Žídek, Augustin; Potapenko, Anna; Bridgland, Alex; Meyer, Clemens; Kohl, Simon A. A.; Ballard, Andrew J.; Cowie, Andrew (August 2021). \"Highly accurate protein structure prediction with AlphaFold\". Nature. 596 (7873): 583–589. Bibcode:2021Natur.596..583J. doi:10.1038/s41586-021-03819-2. ISSN\\xa01476-4687. PMC\\xa08371605. PMID\\xa034265844.\\n\\n^ \"AlphaFold Protein Structure Database\". alphafold.ebi.ac.uk. Archived from the original on 24 July 2021. Retrieved 10 October 2022.\\n\\n^ Titz B, Rajagopala SV, Goll J, Häuser R, McKevitt MT, Palzkill T, Uetz P (May 2008). Hall N (ed.). \"The binary protein interactome of Treponema pallidum--the syphilis spirochete\". PLOS ONE. 3 (5): e2292. Bibcode:2008PLoSO...3.2292T. doi:10.1371/journal.pone.0002292. PMC\\xa02386257. PMID\\xa018509523.\\n\\n^ a b \"Open Bioinformatics Foundation: About us\". Official website. Open Bioinformatics Foundation. Archived from the original on 12 May 2011. Retrieved 10 May 2011.\\n\\n^ \"Open Bioinformatics Foundation: BOSC\". Official website. Open Bioinformatics Foundation. Archived from the original on 18 July 2011. Retrieved 10 May 2011.\\n\\n^ Nisbet R, Elder IV J, Miner G (2009). \"Bioinformatics\". Handbook of Statistical Analysis and Data Mining Applications. Academic Press. p.\\xa0328. ISBN\\xa0978-0080912035.\\n\\n^ Office of the Commissioner. \"Advancing Regulatory Science – Sept. 24–25, 2014 Public Workshop: Next Generation Sequencing Standards\". www.fda.gov. Archived from the original on 14 November 2017. Retrieved 30 November 2017.\\n\\n^ Simonyan V, Goecks J, Mazumder R (2017). \"Biocompute Objects-A Step towards Evaluation and Validation of Biomedical Scientific Computations\". PDA Journal of Pharmaceutical Science and Technology. 71 (2): 136–146. doi:10.5731/pdajpst.2016.006734. PMC\\xa05510742. PMID\\xa027974626.\\n\\n^ Office of the Commissioner. \"Advancing Regulatory Science – Community-based development of HTS standards for validating data and computation and encouraging interoperability\". www.fda.gov. Archived from the original on 26 January 2018. Retrieved 30 November 2017.\\n\\n^ Alterovitz G, Dean D, Goble C, Crusoe MR, Soiland-Reyes S, Bell A, et\\xa0al. (December 2018). \"Enabling precision medicine via standard communication of HTS provenance, analysis, and results\". PLOS Biology. 16 (12): e3000099. doi:10.1371/journal.pbio.3000099. PMC\\xa06338479. PMID\\xa030596645.\\n\\n^ BioCompute Object (BCO) project is a collaborative and community-driven framework to standardize HTS computational data. 1. BCO Specification Document: user manual for understanding and creating B., biocompute-objects, 3 September 2017, archived from the original on 27 June 2018, retrieved 30 November 2017\\n\\n^ Campbell, A. Malcolm (1 June 2003). \"Public Access for Teaching Genomics, Proteomics, and Bioinformatics\". Cell Biology Education. 2 (2): 98–111. doi:10.1187/cbe.03-02-0007. PMC\\xa0162192. PMID\\xa012888845.\\n\\n^ Arenas, Miguel (September 2021). \"General considerations for online teaching practices in bioinformatics in the time of COVID ‐19\". Biochemistry and Molecular Biology Education. 49 (5): 683–684. doi:10.1002/bmb.21558. ISSN\\xa01470-8175. PMC\\xa08426940. PMID\\xa034231941.\\n\\n^ Barker D, Ferrier DE, Holland PW, Mitchell JB, Plaisier H, Ritchie MG, Smart SD (August 2013). \"4273π: bioinformatics education on low cost ARM hardware\". BMC Bioinformatics. 13: 522. doi:10.1186/1471-2105-14-243. PMC\\xa03751261. PMID\\xa023937194.\\n\\n^ Barker D, Alderson RG, McDonagh JL, Plaisier H, Comrie MM, Duncan L, et\\xa0al. (2015). \"University-level practical activities in bioinformatics benefit voluntary groups of pupils in the last 2 years of school\". International Journal of STEM Education. 2 (17). doi:10.1186/s40594-015-0030-z. hdl:10023/7704. S2CID\\xa0256396656.\\n\\n^ McDonagh JL, Barker D, Alderson RG (2016). \"Bringing computational science to the public\". SpringerPlus. 5 (259): 259. doi:10.1186/s40064-016-1856-7. PMC\\xa04775721. PMID\\xa027006868.\\n\\n^ Robson JF, Barker D (October 2015). \"Comparison of the protein-coding gene content of Chlamydia trachomatis and Protochlamydia amoebophila using a Raspberry Pi computer\". BMC Research Notes. 8 (561): 561. doi:10.1186/s13104-015-1476-2. PMC\\xa04604092. PMID\\xa026462790.\\n\\n^ Wreggelsworth KM, Barker D (October 2015). \"A comparison of the protein-coding genomes of two green sulphur bacteria, Chlorobium tepidum TLS and Pelodictyon phaeoclathratiforme BU-1\". BMC Research Notes. 8 (565): 565. doi:10.1186/s13104-015-1535-8. PMC\\xa04606965. PMID\\xa026467441.\\n\\nFurther reading[edit]\\n\\nSehgal et al.\\xa0: Structural, phylogenetic and docking studies of D-amino acid oxidase activator(DAOA ), a candidate schizophrenia gene. Theoretical Biology and Medical Modelling 2013 10 :3.\\n\\nAchuthsankar S Nair Computational Biology & Bioinformatics – A gentle Overview Archived 16 December 2008 at the Wayback Machine, Communications of Computer Society of India, January 2007\\n\\nAluru, Srinivas, ed. Handbook of Computational Molecular Biology. Chapman & Hall/Crc, 2006. ISBN\\xa01-58488-406-1 (Chapman & Hall/Crc Computer and Information Science Series)\\n\\nBaldi, P and Brunak, S, Bioinformatics: The Machine Learning Approach, 2nd edition. MIT Press, 2001. ISBN\\xa00-262-02506-X\\n\\nBarnes, M.R. and Gray, I.C., eds., Bioinformatics for Geneticists, first edition. Wiley, 2003. ISBN\\xa00-470-84394-2\\n\\nBaxevanis, A.D. and Ouellette, B.F.F., eds., Bioinformatics: A Practical Guide to the Analysis of Genes and Proteins, third edition. Wiley, 2005. ISBN\\xa00-471-47878-4\\n\\nBaxevanis, A.D., Petsko, G.A., Stein, L.D., and Stormo, G.D., eds., Current Protocols in Bioinformatics. Wiley, 2007. ISBN\\xa00-471-25093-7\\n\\nCristianini, N. and Hahn, M. Introduction to Computational Genomics Archived 4 January 2009 at the Wayback Machine, Cambridge University Press, 2006. (ISBN\\xa09780521671910 |ISBN\\xa00-521-67191-4)\\n\\nDurbin, R., S. Eddy, A. Krogh and G. Mitchison, Biological sequence analysis. Cambridge University Press, 1998. ISBN\\xa00-521-62971-3\\n\\nGilbert D (September 2004). \"Bioinformatics software resources\". Briefings in Bioinformatics. 5 (3): 300–4. doi:10.1093/bib/5.3.300. PMID\\xa015383216.\\n\\nKeedwell, E., Intelligent Bioinformatics: The Application of Artificial Intelligence Techniques to Bioinformatics Problems. Wiley, 2005. ISBN\\xa00-470-02175-6\\n\\nKohane, et al. Microarrays for an Integrative Genomics. The MIT Press, 2002. ISBN\\xa00-262-11271-X\\n\\nLund, O. et al. Immunological Bioinformatics. The MIT Press, 2005. ISBN\\xa00-262-12280-4\\n\\nPachter, Lior and Sturmfels, Bernd. \"Algebraic Statistics for Computational Biology\" Cambridge University Press, 2005. ISBN\\xa00-521-85700-7\\n\\nPevzner, Pavel A. Computational Molecular Biology: An Algorithmic Approach The MIT Press, 2000. ISBN\\xa00-262-16197-4\\n\\nSoinov, L. Bioinformatics and Pattern Recognition Come Together Archived 10 May 2013 at the Wayback Machine Journal of Pattern Recognition Research (JPRR Archived 8 September 2008 at the Wayback Machine), Vol 1 (1) 2006 p.\\xa037–41\\n\\nStevens, Hallam, Life Out of Sequence: A Data-Driven History of Bioinformatics, Chicago: The University of Chicago Press, 2013, ISBN\\xa09780226080208\\n\\nTisdall, James. \"Beginning Perl for Bioinformatics\" O\\'Reilly, 2001. ISBN\\xa00-596-00080-4\\n\\nCatalyzing Inquiry at the Interface of Computing and Biology (2005) CSTB report Archived 28 January 2007 at the Wayback Machine\\n\\nCalculating the Secrets of Life: Contributions of the Mathematical Sciences and computing to Molecular Biology (1995) Archived 6 July 2008 at the Wayback Machine\\n\\nFoundations of Computational and Systems Biology MIT Course\\n\\nComputational Biology: Genomes, Networks, Evolution Free MIT Course Archived 8 April 2013 at the Wayback Machine\\n\\nExternal links[edit]\\n\\nLibrary resources about  Bioinformatics\\n\\nResources in your library\\nResources in other libraries\\n\\nListen to this article (\\n\\n37 minutes)\\n\\nThis audio file was created from a revision of this article dated 20\\xa0September\\xa02013\\n\\n(2013-09-20), and does not reflect subsequent edits.\\n\\nAudio help\\xa0·\\n\\nMore spoken articles)\\n\\nThe dictionary definition of bioinformatics at Wiktionary\\n\\nLearning materials related to Bioinformatics at Wikiversity\\n\\nMedia related to Bioinformatics at Wikimedia Commons\\n\\nBioinformatics Resource Portal (SIB)\\n\\nv\\n\\nt\\n\\ne\\n\\nBioinformatics\\n\\nSequence databases: GenBank, European Nucleotide Archive, DNA Data Bank of Japan and China National GeneBank\\nSecondary databases: UniProt, database of protein sequences grouping together Swiss-Prot, TrEMBL and Protein Information Resource\\nOther databases: BioNumbers, Protein Data Bank, Ensembl and InterPro\\nSpecialised genomic databases: BOLD, Saccharomyces Genome Database, FlyBase, VectorBase, WormBase, Rat Genome Database, PHI-base, Arabidopsis Information Resource, GISAID and Zebrafish Information Network\\n\\nBLAST\\nBowtie\\nClustal\\nEMBOSS\\nHMMER\\nMUSCLE\\nPANGOLIN\\nSAMtools\\nSOAP suite\\nTopHat\\n\\nServer: ExPASy\\nOntology: Gene Ontology\\nRosalind (education platform)\\n\\nBroad Institute\\nComputational Biology Department (CBD)\\nMicrosoft Research - University of Trento Centre for Computational and Systems Biology (COSBI)\\nDatabase Center for Life Science (DBCLS)\\nDNA Data Bank of Japan (DDBJ)\\nEuropean Bioinformatics Institute (EMBL-EBI)\\nEuropean Molecular Biology Laboratory (EMBL)\\nFlatiron Institute\\nJ. Craig Venter Institute (JCVI)\\nMax Planck Institute of Molecular Cell Biology and Genetics (MPI-CBG)\\nUS National Center for Biotechnology Information (NCBI)\\nJapanese Institute of Genetics\\nNetherlands Bioinformatics Centre (NBIC)\\nPhilippine Genome Center (PGC)\\nScripps Research\\nSwiss Institute of Bioinformatics (SIB)\\nWellcome Sanger Institute\\nWhitehead Institute\\n\\nAfrican Society for Bioinformatics and Computational Biology (ASBCB)\\nAustralia Bioinformatics Resource (EMBL-AR)\\nEuropean Molecular Biology network (EMBnet)\\nInternational Nucleotide Sequence Database Collaboration (INSDC)\\nInternational Society for Biocuration (ISB)\\nInternational Society for Computational Biology (ISCB)\\nStudent Council (ISCB-SC)\\nInstitute of Genomics and Integrative Biology (CSIR-IGIB)\\nJapanese Society for Bioinformatics (JSBi)\\n\\nBasel Computational Biology Conference\\u200e ([BC2])\\nEuropean Conference on Computational Biology (ECCB)\\nIntelligent Systems for Molecular Biology (ISMB)\\nInternational Conference on Bioinformatics (InCoB)\\nInternational Conference on Computational Intelligence Methods for Bioinformatics and Biostatistics (CIBB)\\nISCB Africa ASBCB Conference on Bioinformatics\\nPacific Symposium on Biocomputing (PSB)\\nResearch in Computational Molecular Biology (RECOMB)\\n\\nCRAM format\\nFASTA format\\nFASTQ format\\nNeXML format\\nNexus format\\nPileup format\\nSAM format\\nStockholm format\\nVCF format\\n\\nComputational biology\\nList of biobanks\\nList of biological databases\\nMolecular phylogenetics\\nSequencing\\nSequence database\\nSequence alignment\\n\\nCategory\\n Commons\\n\\nv\\n\\nt\\n\\ne\\n\\nInformation processing\\n\\ninformation processes by function\\nperception\\nattention\\ninfluence\\noperating\\ncommunication\\nreasoning\\nlearning\\nstoring\\ndecision-making\\ninformation processing abstractions\\nevent processing\\nsign processesing\\nsignal processing\\ndata processing\\nstream processing\\nagent processing\\nstate processing\\n\\nnatural\\nnature as information processing\\nhumans as information processing systems\\nsociety as information processing system\\nmixed\\nmixed reality\\nbrain–computer interface\\nphysical computing\\nhuman–computer interaction\\nartificial\\nprocessors and processes\\nbio-inspired computing\\nubiquitous computing\\nartificial brain and mind uploading\\nvirtual reality\\nvirtual world\\n\\nin biology\\ncomputational and systems biology\\ngenetic informatics and cellular computing\\ncomputational neuroscience and neurocomputing\\nin cognitive psychology\\ninformation processing theory\\nmind and intelligence\\ncognitive informatics and neuroinformatics\\nbehavior informatics\\nin computer science\\nneural computation\\ncomputation theory\\nalgorithms and information structures\\ncomputational circuits\\nartificial intelligence\\nin philosophy\\ncomputational theory of mind\\nphilosophy of information\\nphilosophy of artificial intelligence\\ninterdisciplinary\\ninformation theory\\ndecision theory\\nsystems theory\\nother\\ninfosphere\\ninforg\\nDecoding the Universe\\ninformation overload\\n\\nv\\n\\nt\\n\\ne\\n\\nOmics\\n\\nGenomics\\n\\nCognitive genomics\\nComputational genomics\\nComparative genomics\\nFunctional genomics\\nGenome project\\nHuman Genome Project\\nMetagenomics\\nHuman Microbiome Project\\nPangenomics\\nPersonal genomics\\nPopulation genomics\\nSocial genomics\\nStructural genomics\\n\\nBioinformatics\\n\\nBiochip\\nCheminformatics\\nChemogenomics\\nConnectomics\\nHuman Connectome Project\\nEpigenomics\\nHuman Epigenome Project\\nGlycomics\\nImmunomics\\nLipidomics\\nMetabolomics\\nMicrobiomics\\nNutrigenomics\\nPaleopolyploidy\\nPharmacogenetics\\nPharmacogenomics\\nSystems biology\\nToxicogenomics\\nTranscriptomics\\n\\nStructural biology\\n\\nProteomics\\nHuman proteome project\\nCall-map proteomics\\nStructure-based drug design\\nExpression proteomics\\n\\n2-D electrophoresis\\nMass spectrometer\\nElectrospray ionization\\nMatrix-assisted laser desorption ionization\\nMatrix-assisted laser desorption ionization-time of flight mass spectrometer\\nMicrofluidic-based tools\\nIsotope affinity tags\\nChromosome conformation capture\\n\\nDNA Data Bank of Japan (JP)\\nEuropean Molecular Biology Laboratory (EU)\\nNational Institutes of Health (USA)\\nWellcome Sanger Institute (UK)\\n\\nList \\n Category\\n\\nv\\n\\nt\\n\\ne\\n\\nBiology\\n\\nIntroduction (Genetics, Evolution)\\nOutline\\nHistory\\nTimeline\\nIndex\\n\\nBiology\\n\\nOverview\\nScience\\nLife\\nProperties (Adaptation,\\xa0 Energy processing,\\xa0Growth,\\xa0Order,\\xa0Regulation,\\xa0Reproduction,\\xa0Response to environment)\\nHierarchy of life (Atom >\\xa0Molecule >\\xa0Organelle >\\xa0Cell >\\xa0Tissue >\\xa0Organ >\\xa0Organ system >\\xa0Organism >\\xa0Population >\\xa0Community >\\xa0Ecosystem >\\xa0Biosphere)\\nReductionistic\\n Emergent property\\nMechanistic\\nScientific method\\n Theory\\n Law\\nPeer review\\nBiology journals\\nChemical basis\\nAtoms\\nAmino acids\\nCarbohydrates\\nChemical bond\\nChemical element\\nLipids\\nMatter\\nQuantum\\nMolecules\\nMonomer\\nNucleic acids\\nOrganic compounds\\npH\\nPolymer\\nProteins\\nWater\\nCells\\nATP\\nCell cycle\\nCell theory\\nCell signaling\\nCellular respiration\\nEnergy transformation\\nEnzyme\\nEukaryote\\nFermentation\\nMetabolism\\nMeiosis\\nMitosis\\nPhotosynthesis\\nProkaryote\\nGenetics\\nDNA\\nEpigenetics\\nEvolutionary developmental biology\\nGene expression\\nGene regulation\\nGenomes\\nMendelian inheritance\\nPost-transcriptional modification\\nEvolution\\nAdaptation\\nEarliest known life forms\\nFunction\\nGenetic drift\\nGene flow\\nHistory of life\\nMacroevolution\\nMicroevolution\\nMutation\\nNatural selection\\nPhylogenetics\\nSpeciation\\nTaxonomy\\nDiversity\\nArchaea\\nAnimals\\nBacteria\\nBiodiversity\\nFungus\\nInvertebrates\\nPlants\\nProtists\\nVertebrates\\nViruses\\nPlant form  and function\\nEpidermis (botany)\\nFlower\\nGround tissue\\nLeaf\\nPhloem\\nPlant stem\\nRoot\\nShoot\\nVascular plant\\nVascular tissue\\nXylem\\nAnimal form and function\\nBreathing\\nCirculatory system\\nEndocrine system\\nDigestive system\\nHomeostasis\\nImmune system\\nInternal environment\\nMuscular system\\nNervous system\\nReproductive system\\nRespiratory system\\nEcology\\nBiogeochemical cycle\\nBiological interaction\\nBiomass\\nBiomes\\nBiosphere\\nClimate\\nClimate change\\nCommunity\\nConservation\\nEcosystem\\nHabitat\\nniche\\nMicrobiome\\nPopulation dynamics\\nResources\\n\\nResearch  methods\\n\\nLaboratory  techniques\\nGenetic engineering\\nTransformation\\nGel electrophoresis\\nChromatography\\nCentrifugation\\nCell culture\\nDNA sequencing\\nDNA microarray\\nGreen fluorescent protein\\nvector\\nEnzyme assay\\nProtein purification\\nWestern blot\\nNorthern blot\\nSouthern blot\\nRestriction enzyme\\nPolymerase chain reaction\\nTwo-hybrid screening\\nin vivo\\nin vitro\\nin silico\\nField techniques\\nBelt transect\\nmark and recapture\\nspecies discovery curve\\n\\nAgricultural science\\nAnatomy\\nAstrobiology\\nBiochemistry\\nBiogeography\\nBiomechanics\\nBiophysics\\nBioinformatics\\nBiosemiotics\\nBiostatistics\\nBiotechnology\\nBotany\\nCell biology\\nCellular microbiology\\nChronobiology\\nCognitive biology\\nComputational biology\\nConservation biology\\nDevelopmental biology\\nEcology\\nEpidemiology\\nEpigenetics\\nEvolutionary biology\\nEvolutionary developmental biology\\nGenetics\\nGenomics\\nHistology\\nHuman biology\\nImmunology\\nLipidology\\nMarine biology\\nMathematical biology\\nMicrobiology\\nMolecular biology\\nMycology\\nNeuroscience\\nPaleontology\\nParasitology\\nPathology\\nPharmacology\\nPhycology\\nPhysiology\\nReproductive biology\\nSociobiology\\nStructural biology\\nSynthetic biology\\nSystematics\\nSystems biology\\nVirology\\nZoology\\n\\nBiology\\nBotanical terms\\nEcological terms\\nPlant morphology terms\\n\\nBiology portal\\n Category\\n Commons\\n WikiProject\\n\\nv\\n\\nt\\n\\ne\\n\\nBranches of biology\\n\\nAbiogenesis\\nAerobiology\\nAgronomy\\nAgrostology\\nAnatomy\\nAstrobiology\\nBacteriology\\nBiochemistry\\nBiogeography\\nBiogeology\\nBioinformatics\\nBiological engineering\\nBiomechanics\\nBiophysics\\nBiosemiotics\\nBiostatistics\\nBiotechnology\\nBotany\\nCell biology\\nCellular microbiology\\nChemical biology\\nChronobiology\\nCognitive biology\\nComputational biology\\nConservation biology\\nCryobiology\\nCytogenetics\\nDendrology\\nDevelopmental biology\\nEcological genetics\\nEcology\\nEmbryology\\nEpidemiology\\nEpigenetics\\nEvolutionary biology\\nFreshwater biology\\nGenetics\\nGenomics\\nGeobiology\\nGerontology\\nHerpetology\\nHistology\\nHuman biology\\nIchthyology\\nImmunology\\nLipidology\\nMammalogy\\nMarine biology\\nMathematical biology\\nMicrobiology\\nMolecular biology\\nMycology\\nNeontology\\nNeuroscience\\nNutrition\\nOrnithology\\nOsteology\\nPaleontology\\nParasitology\\nPathology\\nPharmacology\\nPhotobiology\\nPhycology\\nPhylogenetics\\nPhysiology\\nPomology\\nPrimatology\\nProteomics\\nProtistology\\nQuantum biology\\nReproductive biology\\nSociobiology\\nStructural biology\\nSynthetic biology\\nSystematics\\nSystems biology\\nTaxonomy\\nTeratology\\nToxicology\\nVirology\\nVirophysics\\nXenobiology\\nZoology\\n\\nHistory of biology\\nNobel Prize in Physiology or Medicine\\nTimeline of biology and organic chemistry\\n\\nv\\n\\nt\\n\\ne\\n\\nComputer science\\n\\nNote: This template roughly follows the 2012 ACM Computing Classification System.\\n\\nHardware\\n\\nPrinted circuit board\\nPeripheral\\nIntegrated circuit\\nVery Large Scale Integration\\nSystems on Chip (SoCs)\\nEnergy consumption (Green computing)\\nElectronic design automation\\nHardware acceleration\\n\\nComputer architecture\\nEmbedded system\\nReal-time computing\\nDependability\\n\\nNetworks\\n\\nNetwork architecture\\nNetwork protocol\\nNetwork components\\nNetwork scheduler\\nNetwork performance evaluation\\nNetwork service\\n\\nInterpreter\\nMiddleware\\nVirtual machine\\nOperating system\\nSoftware quality\\n\\nSoftware notations and\\n\\ntools\\n\\nProgramming paradigm\\nProgramming language\\nCompiler\\nDomain-specific language\\nModeling language\\nSoftware framework\\nIntegrated development environment\\nSoftware configuration management\\nSoftware library\\nSoftware repository\\n\\nSoftware development\\n\\nControl variable\\nSoftware development process\\nRequirements analysis\\nSoftware design\\nSoftware construction\\nSoftware deployment\\nSoftware engineering\\nSoftware maintenance\\nProgramming team\\nOpen-source model\\n\\nTheory of computation\\n\\nModel of computation\\nFormal language\\nAutomata theory\\nComputability theory\\nComputational complexity theory\\nLogic\\nSemantics\\n\\nAlgorithms\\n\\nAlgorithm design\\nAnalysis of algorithms\\nAlgorithmic efficiency\\nRandomized algorithm\\nComputational geometry\\n\\nDiscrete mathematics\\nProbability\\nStatistics\\nMathematical software\\nInformation theory\\nMathematical analysis\\nNumerical analysis\\nTheoretical computer science\\n\\nInformation systems\\n\\nDatabase management system\\nInformation storage systems\\nEnterprise information system\\nSocial information systems\\nGeographic information system\\nDecision support system\\nProcess control system\\nMultimedia information system\\nData mining\\nDigital library\\nComputing platform\\nDigital marketing\\nWorld Wide Web\\nInformation retrieval\\n\\nSecurity\\n\\nCryptography\\nFormal methods\\nSecurity services\\nIntrusion detection system\\nHardware security\\nNetwork security\\nInformation security\\nApplication security\\n\\nHuman–computer interaction\\n\\nInteraction design\\nSocial computing\\nUbiquitous computing\\nVisualization\\nAccessibility\\n\\nConcurrency\\n\\nConcurrent computing\\nParallel computing\\nDistributed computing\\nMultithreading\\nMultiprocessing\\n\\nArtificial intelligence\\n\\nNatural language processing\\nKnowledge representation and reasoning\\nComputer vision\\nAutomated planning and scheduling\\nSearch methodology\\nControl method\\nPhilosophy of artificial intelligence\\nDistributed artificial intelligence\\n\\nMachine learning\\n\\nSupervised learning\\nUnsupervised learning\\nReinforcement learning\\nMulti-task learning\\nCross-validation\\n\\nGraphics\\n\\nAnimation\\nRendering\\nPhotograph manipulation\\nGraphics processing unit\\nMixed reality\\nVirtual reality\\nImage compression\\nSolid modeling\\n\\nE-commerce\\nEnterprise software\\nComputational mathematics\\nComputational physics\\nComputational chemistry\\nComputational biology\\nComputational social science\\nComputational engineering\\nComputational healthcare\\nDigital art\\nElectronic publishing\\nCyberwarfare\\nElectronic voting\\nVideo games\\nWord processing\\nOperations research\\nEducational technology\\nDocument management\\n\\nCategory\\n Outline\\nWikiProject\\n Commons\\n\\nv\\n\\nt\\n\\ne\\n\\nHealth informatics\\n\\nMedical image computing and imaging informatics\\nArtificial intelligence in healthcare\\nNeuroinformatics in healthcare\\nBehavior informatics in healthcare\\nComputational biology in healthcare\\nTranslational bioinformatics\\nTranslational medicine\\nhealth information technology\\nTelemedicine\\nPublic health informatics\\nHealth information management\\nConsumer health informatics\\n\\nMedical classification\\n\\nContinuity of Care Record\\nHRHIS\\nICD\\nISO 27799\\nLOINC\\n\\nAmerican Association for Medical Systems and Informatics\\nAmerican Medical Informatics Association\\nAustralian Society / Australasian College (to merge from 2020)\\nBrazilian Society\\nEuropean Federation\\nIndian Association\\nInternational Association\\nAmerican College of Medical Informatics\\n\\nElectronic health record\\nHealth Level 7\\nRemote manipulator\\nPersonalized medicine / precision medicine\\nList of medical and health informatics journals\\nopenEHR\\n\\nPortals:\\n\\nBiology\\n\\nEvolutionary biology\\n\\nAuthority control: National\\n\\nFrance\\nBnF data\\nGermany\\nIsrael\\nUnited States\\nJapan\\nCzech Republic\\n\\nRetrieved from \"\\n\\nhttps://en.wikipedia.org/w/index.php?title=Bioinformatics&oldid=1170322562\"\\n\\nCategory:\\n\\nBioinformatics\\n\\nHidden categories: \\n\\nArticles with short description\\n\\nShort description matches Wikidata\\n\\nUse dmy dates from September 2020\\n\\nArticles with hAudio microformats\\n\\nPages including recorded pronunciations\\n\\nAll articles with vague or ambiguous time\\n\\nVague or ambiguous time from June 2023\\n\\nAll articles with specifically marked weasel-worded phrases\\n\\nArticles with specifically marked weasel-worded phrases from June 2020\\n\\nAll articles with unsourced statements\\n\\nArticles with unsourced statements from June 2023\\n\\nArticles with unsourced statements from July 2015\\n\\nWikipedia articles needing clarification from June 2023\\n\\nWebarchive template wayback links\\n\\nSpoken articles\\n\\nCommons category link from Wikidata\\n\\nArticles with BNF identifiers\\n\\nArticles with BNFdata identifiers\\n\\nArticles with GND identifiers\\n\\nArticles with J9U identifiers\\n\\nArticles with LCCN identifiers\\n\\nArticles with NDL identifiers\\n\\nArticles with NKC identifiers', metadata={'source': 'https://en.wikipedia.org/wiki/Bioinformatics'})"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
        "\n",
        "def split_docs(documents):\n",
        "  text_splitter = RecursiveCharacterTextSplitter(\n",
        "      # Set a really small chunk size, just to show.\n",
        "      chunk_size = 3000,\n",
        "      chunk_overlap  = 500\n",
        "  )\n",
        "  docs = text_splitter.split_documents(documents)\n",
        "\n",
        "  return docs"
      ],
      "metadata": {
        "id": "awpR7R_RKVN5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "docs = split_docs(documents)"
      ],
      "metadata": {
        "id": "1vUEaqiUKVLP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "len(docs)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WJKYIkNrKVJC",
        "outputId": "80906b1b-5a6d-435a-92d6-d2299c67a1b3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "6772"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(docs[0].page_content)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GV8vkZagvjOA",
        "outputId": "0b2e2dd1-af17-4e34-b541-e91e8d8f059a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Toggle the table of contents\n",
            "\n",
            "Toggle the table of contents\n",
            "\n",
            "Bioinformatics\n",
            "\n",
            "76 languages\n",
            "\n",
            "العربية\n",
            "\n",
            "Azərbaycanca\n",
            "\n",
            "বাংলা\n",
            "\n",
            "Bân-lâm-gú\n",
            "\n",
            "Беларуская\n",
            "\n",
            "Български\n",
            "\n",
            "Bosanski\n",
            "\n",
            "Català\n",
            "\n",
            "Čeština\n",
            "\n",
            "Dansk\n",
            "\n",
            "Deutsch\n",
            "\n",
            "Eesti\n",
            "\n",
            "Ελληνικά\n",
            "\n",
            "Español\n",
            "\n",
            "Esperanto\n",
            "\n",
            "Euskara\n",
            "\n",
            "فارسی\n",
            "\n",
            "Français\n",
            "\n",
            "Galego\n",
            "\n",
            "Gĩkũyũ\n",
            "\n",
            "한국어\n",
            "\n",
            "Հայերեն\n",
            "\n",
            "हिन्दी\n",
            "\n",
            "Hrvatski\n",
            "\n",
            "Ido\n",
            "\n",
            "Bahasa Indonesia\n",
            "\n",
            "Íslenska\n",
            "\n",
            "Italiano\n",
            "\n",
            "עברית\n",
            "\n",
            "Jawa\n",
            "\n",
            "ಕನ್ನಡ\n",
            "\n",
            "Қазақша\n",
            "\n",
            "Latina\n",
            "\n",
            "Latviešu\n",
            "\n",
            "Lëtzebuergesch\n",
            "\n",
            "Limburgs\n",
            "\n",
            "Lingua Franca Nova\n",
            "\n",
            "Magyar\n",
            "\n",
            "Македонски\n",
            "\n",
            "മലയാളം\n",
            "\n",
            "მარგალური\n",
            "\n",
            "مصرى\n",
            "\n",
            "Bahasa Melayu\n",
            "\n",
            "Монгол\n",
            "\n",
            "Nederlands\n",
            "\n",
            "日本語\n",
            "\n",
            "Norsk bokmål\n",
            "\n",
            "Norsk nynorsk\n",
            "\n",
            "Novial\n",
            "\n",
            "Occitan\n",
            "\n",
            "پښتو\n",
            "\n",
            "Polski\n",
            "\n",
            "Português\n",
            "\n",
            "Română\n",
            "\n",
            "Русский\n",
            "\n",
            "Shqip\n",
            "\n",
            "Simple English\n",
            "\n",
            "Slovenčina\n",
            "\n",
            "Slovenščina\n",
            "\n",
            "کوردی\n",
            "\n",
            "Српски / srpski\n",
            "\n",
            "Srpskohrvatski / српскохрватски\n",
            "\n",
            "Suomi\n",
            "\n",
            "Svenska\n",
            "\n",
            "Tagalog\n",
            "\n",
            "தமிழ்\n",
            "\n",
            "ไทย\n",
            "\n",
            "Türkçe\n",
            "\n",
            "Українська\n",
            "\n",
            "اردو\n",
            "\n",
            "Tiếng Việt\n",
            "\n",
            "Winaray\n",
            "\n",
            "吴语\n",
            "\n",
            "ייִדיש\n",
            "\n",
            "粵語\n",
            "\n",
            "中文\n",
            "\n",
            "Edit links\n",
            "\n",
            "Article\n",
            "\n",
            "Talk\n",
            "\n",
            "English\n",
            "\n",
            "Read\n",
            "\n",
            "Edit\n",
            "\n",
            "View history\n",
            "\n",
            "Tools\n",
            "\n",
            "Tools\n",
            "\n",
            "Actions\n",
            "\n",
            "Read\n",
            "\n",
            "Edit\n",
            "\n",
            "View history\n",
            "\n",
            "General\n",
            "\n",
            "What links here\n",
            "\n",
            "Related changes\n",
            "\n",
            "Upload file\n",
            "\n",
            "Special pages\n",
            "\n",
            "Permanent link\n",
            "\n",
            "Page information\n",
            "\n",
            "Cite this page\n",
            "\n",
            "Wikidata item\n",
            "\n",
            "Print/export\n",
            "\n",
            "Download as PDF\n",
            "\n",
            "Printable version\n",
            "\n",
            "In other projects\n",
            "\n",
            "Wikimedia Commons\n",
            "\n",
            "Wikibooks\n",
            "\n",
            "Wikiversity\n",
            "\n",
            "From Wikipedia, the free encyclopedia\n",
            "\n",
            "Computational analysis of large, complex sets of biological data\n",
            "\n",
            "For the journal, see \n",
            "\n",
            "Bioinformatics (journal).\n",
            "\n",
            "§ Sequence analysis for further information.\n",
            "\n",
            "National Center for Biotechnology Information website)\n",
            "\n",
            "Not to be confused with \n",
            "\n",
            "Biological computation or\n",
            "\n",
            "Genetic algorithm.\n",
            "\n",
            "/ˌbaɪ.oʊˌɪnfərˈmætɪks/ (listen)) is an\n",
            "\n",
            "interdisciplinary field of\n",
            "\n",
            "science that develops methods and\n",
            "\n",
            "software tools for understanding\n",
            "\n",
            "biological data, especially when the data sets are large and complex. Bioinformatics uses\n",
            "\n",
            "biology,\n",
            "\n",
            "chemistry,\n",
            "\n",
            "physics,\n",
            "\n",
            "computer science,\n",
            "\n",
            "computer programming,\n",
            "\n",
            "information engineering,\n",
            "\n",
            "mathematics and\n",
            "\n",
            "statistics to analyze and interpret\n",
            "\n",
            "biological data. The subsequent process of analyzing and interpreting data is referred to as\n",
            "\n",
            "computational biology.\n",
            "\n",
            "computer simulation analyses of biological queries. They include reused specific analysis \"pipelines\", particularly in the field of\n",
            "\n",
            "genomics, such as by the identification of\n",
            "\n",
            "genes and single\n",
            "\n",
            "nucleotide polymorphisms (\n",
            "\n",
            "SNPs). These pipelines are used to better understand the genetic basis of disease, unique adaptations, desirable properties (esp. in agricultural species), or differences between populations. Bioinformatics also includes\n",
            "\n",
            "proteomics, which tries to understand the organizational principles within\n",
            "\n",
            "nucleic acid and\n",
            "\n",
            "protein sequences.\n",
            "\n",
            "[1]\n",
            "\n",
            "signal processing allow extraction of useful results from large amounts of raw data. In the field of genetics, it aids in sequencing and annotating genomes and their observed\n",
            "\n",
            "mutations. Bioinformatics includes\n",
            "\n",
            "text mining of biological literature and the development of biological and gene\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install tiktoken -q"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MoJ78-b15nxE",
        "outputId": "dad7e631-8509-4bdb-e730-5ce8976b7d0a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/1.7 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[91m━━\u001b[0m\u001b[91m╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.1/1.7 MB\u001b[0m \u001b[31m3.2 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━\u001b[0m\u001b[90m╺\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.4/1.7 MB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.8/1.7 MB\u001b[0m \u001b[31m7.3 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[90m╺\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.7 MB\u001b[0m \u001b[31m9.1 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m10.3 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m9.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install 'google-cloud-aiplatform>=1.25.0'"
      ],
      "metadata": {
        "id": "U1125TP1_yRx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import openai"
      ],
      "metadata": {
        "id": "XOHHrJSdwJ4X"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from langchain.embeddings.openai import OpenAIEmbeddings\n",
        "\n",
        "\n",
        "embeddings = OpenAIEmbeddings()\n",
        "\n",
        "query_result = embeddings.embed_query(\"Hello world\")\n",
        "len(query_result)\n",
        "\n",
        "# print(query_result)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9fGEhSkl1zZa",
        "outputId": "bfa3d737-82c0-4403-df4d-14581e2ddb74"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "1536"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install pinecone-client -q\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ee8CyTHX1zXJ",
        "outputId": "c4c4f2f8-721e-48dc-dca8-3d19cd0c82a5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m179.1/179.1 kB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m60.0/60.0 kB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m300.4/300.4 kB\u001b[0m \u001b[31m6.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd"
      ],
      "metadata": {
        "id": "I6dETZ8fkjLg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df = pd.DataFrame(docs, columns =['text', 'source'])\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "xTU6CRdwkfgV",
        "outputId": "810ed0f6-20d2-4b2d-ea63-71b5e4bd1595"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   text  \\\n",
              "0     (page_content, Toggle the table of contents\\n\\...   \n",
              "1     (page_content, proteomics, which tries to unde...   \n",
              "2     (page_content, [18]\\n\\nIn the 1970s, new techn...   \n",
              "3     (page_content, Sequence analysis[edit]\\n\\nMain...   \n",
              "4     (page_content, See also: \\n\\nsequence analysis...   \n",
              "...                                                 ...   \n",
              "6766  (page_content, Applications[edit]\\n\\nDetermini...   \n",
              "6767  (page_content, Identifying genes in biological...   \n",
              "6768  (page_content, See also[edit]\\n\\nChemical biol...   \n",
              "6769  (page_content, ^ a b Wuster A, Madan Babu M (M...   \n",
              "6770  (page_content, Bioinformatics\\n\\nBiochip\\nChem...   \n",
              "\n",
              "                                                 source  \n",
              "0     (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "1     (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "2     (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "3     (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "4     (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "...                                                 ...  \n",
              "6766  (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "6767  (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "6768  (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "6769  (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "6770  (metadata, {'source': 'https://en.wikipedia.or...  \n",
              "\n",
              "[6771 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "\n",
              "  <div id=\"df-abcb1aab-e611-4063-b54c-f77f657022f8\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>source</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>(page_content, Toggle the table of contents\\n\\...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>(page_content, proteomics, which tries to unde...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>(page_content, [18]\\n\\nIn the 1970s, new techn...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>(page_content, Sequence analysis[edit]\\n\\nMain...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>(page_content, See also: \\n\\nsequence analysis...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6766</th>\n",
              "      <td>(page_content, Applications[edit]\\n\\nDetermini...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6767</th>\n",
              "      <td>(page_content, Identifying genes in biological...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6768</th>\n",
              "      <td>(page_content, See also[edit]\\n\\nChemical biol...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6769</th>\n",
              "      <td>(page_content, ^ a b Wuster A, Madan Babu M (M...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6770</th>\n",
              "      <td>(page_content, Bioinformatics\\n\\nBiochip\\nChem...</td>\n",
              "      <td>(metadata, {'source': 'https://en.wikipedia.or...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6771 rows × 2 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-abcb1aab-e611-4063-b54c-f77f657022f8')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "\n",
              "\n",
              "\n",
              "    <div id=\"df-f0942f64-51e9-43ec-b023-93ef15f4f346\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f0942f64-51e9-43ec-b023-93ef15f4f346')\"\n",
              "              title=\"Suggest charts.\"\n",
              "              style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "    </div>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "    background-color: #E8F0FE;\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: #1967D2;\n",
              "    height: 32px;\n",
              "    padding: 0 0 0 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: #E2EBFA;\n",
              "    box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: #174EA6;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "    background-color: #3B4455;\n",
              "    fill: #D2E3FC;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart:hover {\n",
              "    background-color: #434B5C;\n",
              "    box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "    filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "    fill: #FFFFFF;\n",
              "  }\n",
              "</style>\n",
              "\n",
              "    <script>\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#' + key);\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      }\n",
              "    </script>\n",
              "\n",
              "      <script>\n",
              "\n",
              "function displayQuickchartButton(domScope) {\n",
              "  let quickchartButtonEl =\n",
              "    domScope.querySelector('#df-f0942f64-51e9-43ec-b023-93ef15f4f346 button.colab-df-quickchart');\n",
              "  quickchartButtonEl.style.display =\n",
              "    google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "}\n",
              "\n",
              "        displayQuickchartButton(document);\n",
              "      </script>\n",
              "      <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-abcb1aab-e611-4063-b54c-f77f657022f8 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-abcb1aab-e611-4063-b54c-f77f657022f8');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 42
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df['text'][0]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Yev6mXeDlfYg",
        "outputId": "7b2cb6b3-5328-42fc-e77e-4c3cc83ba5b5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "('page_content',\n",
              " 'Toggle the table of contents\\n\\nToggle the table of contents\\n\\nBioinformatics\\n\\n76 languages\\n\\nالعربية\\n\\nAzərbaycanca\\n\\nবাংলা\\n\\nBân-lâm-gú\\n\\nБеларуская\\n\\nБългарски\\n\\nBosanski\\n\\nCatalà\\n\\nČeština\\n\\nDansk\\n\\nDeutsch\\n\\nEesti\\n\\nΕλληνικά\\n\\nEspañol\\n\\nEsperanto\\n\\nEuskara\\n\\nفارسی\\n\\nFrançais\\n\\nGalego\\n\\nGĩkũyũ\\n\\n한국어\\n\\nՀայերեն\\n\\nहिन्दी\\n\\nHrvatski\\n\\nIdo\\n\\nBahasa Indonesia\\n\\nÍslenska\\n\\nItaliano\\n\\nעברית\\n\\nJawa\\n\\nಕನ್ನಡ\\n\\nҚазақша\\n\\nLatina\\n\\nLatviešu\\n\\nLëtzebuergesch\\n\\nLimburgs\\n\\nLingua Franca Nova\\n\\nMagyar\\n\\nМакедонски\\n\\nമലയാളം\\n\\nმარგალური\\n\\nمصرى\\n\\nBahasa Melayu\\n\\nМонгол\\n\\nNederlands\\n\\n日本語\\n\\nNorsk bokmål\\n\\nNorsk nynorsk\\n\\nNovial\\n\\nOccitan\\n\\nپښتو\\n\\nPolski\\n\\nPortuguês\\n\\nRomână\\n\\nРусский\\n\\nShqip\\n\\nSimple English\\n\\nSlovenčina\\n\\nSlovenščina\\n\\nکوردی\\n\\nСрпски / srpski\\n\\nSrpskohrvatski / српскохрватски\\n\\nSuomi\\n\\nSvenska\\n\\nTagalog\\n\\nதமிழ்\\n\\nไทย\\n\\nTürkçe\\n\\nУкраїнська\\n\\nاردو\\n\\nTiếng Việt\\n\\nWinaray\\n\\n吴语\\n\\nייִדיש\\n\\n粵語\\n\\n中文\\n\\nEdit links\\n\\nArticle\\n\\nTalk\\n\\nEnglish\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nTools\\n\\nTools\\n\\nActions\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nGeneral\\n\\nWhat links here\\n\\nRelated changes\\n\\nUpload file\\n\\nSpecial pages\\n\\nPermanent link\\n\\nPage information\\n\\nCite this page\\n\\nWikidata item\\n\\nPrint/export\\n\\nDownload as PDF\\n\\nPrintable version\\n\\nIn other projects\\n\\nWikimedia Commons\\n\\nWikibooks\\n\\nWikiversity\\n\\nFrom Wikipedia, the free encyclopedia\\n\\nComputational analysis of large, complex sets of biological data\\n\\nFor the journal, see \\n\\nBioinformatics (journal).\\n\\n§\\xa0Sequence analysis for further information.\\n\\nNational Center for Biotechnology Information website)\\n\\nNot to be confused with \\n\\nBiological computation or\\n\\nGenetic algorithm.\\n\\n/ˌbaɪ.oʊˌɪnfərˈmætɪks/ (listen)) is an\\n\\ninterdisciplinary field of\\n\\nscience that develops methods and\\n\\nsoftware tools for understanding\\n\\nbiological data, especially when the data sets are large and complex. Bioinformatics uses\\n\\nbiology,\\n\\nchemistry,\\n\\nphysics,\\n\\ncomputer science,\\n\\ncomputer programming,\\n\\ninformation engineering,\\n\\nmathematics and\\n\\nstatistics to analyze and interpret\\n\\nbiological data. The subsequent process of analyzing and interpreting data is referred to as\\n\\ncomputational biology.\\n\\ncomputer simulation analyses of biological queries. They include reused specific analysis \"pipelines\", particularly in the field of\\n\\ngenomics, such as by the identification of\\n\\ngenes and single\\n\\nnucleotide polymorphisms (\\n\\nSNPs). These pipelines are used to better understand the genetic basis of disease, unique adaptations, desirable properties (esp. in agricultural species), or differences between populations. Bioinformatics also includes\\n\\nproteomics, which tries to understand the organizational principles within\\n\\nnucleic acid and\\n\\nprotein sequences.\\n\\n[1]\\n\\nsignal processing allow extraction of useful results from large amounts of raw data. In the field of genetics, it aids in sequencing and annotating genomes and their observed\\n\\nmutations. Bioinformatics includes\\n\\ntext mining of biological literature and the development of biological and gene')"
            ]
          },
          "metadata": {},
          "execution_count": 48
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pinecone\n",
        "from langchain.vectorstores import Pinecone\n",
        "\n",
        "pinecone.init(\n",
        "    api_key=\"3a8c2abe-4851-4dab-997f-85ace69a0d28\",\n",
        "    environment=\"gcp-starter\"\n",
        ")\n",
        "\n",
        "index_name = \"bioinformatics-chatbot-llm\"\n",
        "\n",
        "# index = Pinecone.from_documents(docs[:1000], embeddings, index_name=index_name)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xgAtZsGO1zVc",
        "outputId": "7da2c896-cf5e-404e-bd9c-a1aaff1bcc96"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/pinecone/index.py:4: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)\n",
            "  from tqdm.autonotebook import tqdm\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "index = pinecone.Index('bioinformatics-chatbot-llm')"
      ],
      "metadata": {
        "id": "wQj4Y8Hu59wj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pinecone.list_indexes()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9QzdyHMQ6WwE",
        "outputId": "497be317-1495-43ea-b40b-89dbd213a676"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['bioinformatics-chatbot-llm']"
            ]
          },
          "metadata": {},
          "execution_count": 51
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "index = Pinecone.from_documents(docs[128:256], embeddings, index_name=index_name)"
      ],
      "metadata": {
        "id": "TqBO9PLk1zS1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import time\n",
        "import pandas as pd\n",
        "import json\n"
      ],
      "metadata": {
        "id": "feLgy9Jm0KFA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def handle_quota_errors(func, *args, retry_delay=5,backoff_factor=2, **kwargs):\n",
        "    retries = 0\n",
        "\n",
        "    try:\n",
        "      time.sleep(10)\n",
        "      return func(*args, **kwargs)\n",
        "    except Exception as e:\n",
        "      print(f\"error: {e}\")\n",
        "      retries += 1\n",
        "      wait = retry_delay * (backoff_factor ** retries)\n",
        "      time.sleep(wait)\n",
        "      print(f\"wait for {wait} seconds\")\n",
        "\n",
        "index = handle_quota_errors(Pinecone.from_documents, docs[1152:], embeddings, index_name=index_name)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ezcdWli_kDiL",
        "outputId": "3745ac6a-02a9-45f0-99fd-5e2eb28da6c7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per day. Limit: 200 / day. Please try again in 7m12s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method..\n",
            "WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per day. Limit: 200 / day. Please try again in 7m12s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method..\n",
            "WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per min. Limit: 3 / min. Please try again in 20s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method..\n",
            "WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 8.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per min. Limit: 3 / min. Please try again in 20s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method..\n",
            "WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 10.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per day. Limit: 200 / day. Please try again in 7m12s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method..\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "error: Rate limit reached for default-text-embedding-ada-002 in organization org-m1qKM7h9GXBIzihGlD5WLgke on requests per min. Limit: 3 / min. Please try again in 20s. Contact us through our help center at help.openai.com if you continue to have issues. Please add a payment method to your account to increase your rate limit. Visit https://platform.openai.com/account/billing to add a payment method.\n",
            "wait for 10 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for i in range(len(docs)-1217):\n",
        "  idx = 1217 + i\n",
        "  index = Pinecone.from_documents(docs[idx:idx+1], embeddings, index_name=index_name)\n",
        "  time.sleep(10)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 228
        },
        "id": "GNsuomuV1zOU",
        "outputId": "a57b7cf7-8ff9-4806-baf4-c756d32659c2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-34-bdbb911761e0>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdocs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1217\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m   \u001b[0midx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1217\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m   \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPinecone\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_documents\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdocs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0membeddings\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindex_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m   \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'Pinecone' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def get_similiar_docs(query_vec, k=5, score=False):\n",
        "  if score:\n",
        "    similar_docs = index.similarity_search_with_score(query, k=k)\n",
        "  else:\n",
        "    similar_docs = idx.similarity_search(query, k=k)\n",
        "  return similar_docs\n"
      ],
      "metadata": {
        "id": "2KTeYB-YioGM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "query_vec = embeddings.embed_query(query)\n",
        "simi_docs = get_similiar_docs(query_vec, index)\n",
        "simi_docs"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MqDf-4yHqCk8",
        "outputId": "83b4eae7-2b1e-44fe-8372-4e905a4fb9b7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'matches': [{'id': '9b29c531-e84c-4e18-bb69-e82ecb18eb92',\n",
              "              'score': 0.839030385,\n",
              "              'values': []},\n",
              "             {'id': 'fbb8e443-2402-47e1-9168-d90f094e79ff',\n",
              "              'score': 0.834465325,\n",
              "              'values': []},\n",
              "             {'id': 'da2be476-235d-4d01-ad79-1b283f1c9375',\n",
              "              'score': 0.833740473,\n",
              "              'values': []},\n",
              "             {'id': '6d5e848e-63c9-4a98-9e0d-0c1d711f4924',\n",
              "              'score': 0.833513498,\n",
              "              'values': []},\n",
              "             {'id': '5ef2d469-aa1c-4416-9857-ca9388076d19',\n",
              "              'score': 0.829659104,\n",
              "              'values': []}],\n",
              " 'namespace': ''}"
            ]
          },
          "metadata": {},
          "execution_count": 60
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "index.query(id=\"9b29c531-e84c-4e18-bb69-e82ecb18eb92\", top_k=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7opKtt9NyqDH",
        "outputId": "5fd260e5-7f6c-4dba-dc4a-156365316075"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'matches': [{'id': '9b29c531-e84c-4e18-bb69-e82ecb18eb92',\n",
              "              'score': 1.00509715,\n",
              "              'values': []}],\n",
              " 'namespace': ''}"
            ]
          },
          "metadata": {},
          "execution_count": 62
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "idx = Pinecone.from_existing_index(index_name='bioinformatics-chatbot-llm', embedding=embeddings)"
      ],
      "metadata": {
        "id": "gkuExKvvz2aw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "idx.similarity_search(query, k=5)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pMQ9Muaw1rKU",
        "outputId": "c8f4cfd3-bf96-4319-d1dc-0cfbdae56ec5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[Document(page_content='Toggle the table of contents\\n\\nToggle the table of contents\\n\\nGenomics\\n\\n50 languages\\n\\nالعربية\\n\\nAzərbaycanca\\n\\nবাংলা\\n\\nBân-lâm-gú\\n\\nБашҡортса\\n\\nБеларуская\\n\\nBosanski\\n\\nCatalà\\n\\nČeština\\n\\nEesti\\n\\nΕλληνικά\\n\\nEspañol\\n\\nEsperanto\\n\\nEuskara\\n\\nفارسی\\n\\nFrançais\\n\\nGaeilge\\n\\nGalego\\n\\n한국어\\n\\nՀայերեն\\n\\nहिन्दी\\n\\nBahasa Indonesia\\n\\nÍslenska\\n\\nItaliano\\n\\nעברית\\n\\nҚазақша\\n\\nLietuvių\\n\\nMagyar\\n\\nBahasa Melayu\\n\\nNederlands\\n\\n日本語\\n\\nNorsk bokmål\\n\\nNovial\\n\\nOʻzbekcha / ўзбекча\\n\\nپښتو\\n\\nPolski\\n\\nPortuguês\\n\\nRomână\\n\\nРусский\\n\\nSimple English\\n\\nСрпски / srpski\\n\\nSrpskohrvatski / српскохрватски\\n\\nSuomi\\n\\nSvenska\\n\\nதமிழ்\\n\\nไทย\\n\\nTürkçe\\n\\nУкраїнська\\n\\nTiếng Việt\\n\\n中文\\n\\nEdit links\\n\\nArticle\\n\\nTalk\\n\\nEnglish\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nTools\\n\\nTools\\n\\nActions\\n\\nRead\\n\\nEdit\\n\\nView history\\n\\nGeneral\\n\\nWhat links here\\n\\nRelated changes\\n\\nUpload file\\n\\nSpecial pages\\n\\nPermanent link\\n\\nPage information\\n\\nCite this page\\n\\nWikidata item\\n\\nPrint/export\\n\\nDownload as PDF\\n\\nPrintable version\\n\\nIn other projects\\n\\nWikimedia Commons\\n\\nWikiquote\\n\\nFrom Wikipedia, the free encyclopedia\\n\\nDiscipline in genetics\\n\\nThis article is about the scientific field. For the journal, see \\n\\nGenomics (journal).\\n\\n\"Genome biology\" redirects here. For the journal with the same name, see \\n\\nGenome Biology.\\n\\nPart of a series on\\n\\nGenetics\\n\\nKey components\\nChromosome\\nDNA\\nRNA\\nGenome\\nHeredity\\nMutation\\nNucleotide\\nVariation\\n\\nOutline\\nIndex\\n\\nHistory and topics\\nIntroduction\\nHistory\\nEvolution (molecular)\\nPopulation genetics\\nMendelian inheritance\\nQuantitative genetics\\nMolecular genetics\\n\\nResearch\\nGeneticist\\nDNA sequencing\\nGenetic engineering\\nGenomics ( template)\\nMedical genetics\\n\\nBranches of genetics\\n\\nPersonalized medicine\\nPersonalized medicine\\n\\nCategory\\n\\n.mw-parser-output .navbar{display:inline;font-size:88%;font-weight:normal}.mw-parser-output .navbar-collapse{float:left;text-align:left}.mw-parser-output .navbar-boxtext{word-spacing:0}.mw-parser-output .navbar ul{display:inline-block;white-space:nowrap;line-height:inherit}.mw-parser-output .navbar-brackets::before{margin-right:-0.125em;content:\"[ \"}.mw-parser-output .navbar-brackets::after{margin-left:-0.125em;content:\" ]\"}.mw-parser-output .navbar li{word-spacing:-0.125em}.mw-parser-output .navbar a>span,.mw-parser-output .navbar a>abbr{text-decoration:inherit}.mw-parser-output .navbar-mini abbr{font-variant:small-caps;border-bottom:none;text-decoration:none;cursor:inherit}.mw-parser-output .navbar-ct-full{font-size:114%;margin:0 7em}.mw-parser-output .navbar-ct-mini{font-size:114%;margin:0 4em}vte\\n\\nbiology focusing on the structure, function, evolution, mapping, and editing of\\n\\ngenomes. A genome is an organism\\'s complete set of\\n\\nDNA, including all of its genes as well as its hierarchical, three-dimensional structural configuration.\\n\\n[1]\\n\\n[2]\\n\\n[3]\\n\\n[4]\\n\\ngenetics, which refers to the study of\\n\\n[5]', metadata={'source': 'https://en.wikipedia.org/wiki/Genomics'}),\n",
              " Document(page_content='Research areas[edit]\\n\\nFunctional genomics[edit]\\n\\nMain article: \\n\\nFunctional genomics\\n\\nmolecular biology that attempts to make use of the vast wealth of data produced by genomic projects (such as\\n\\ngenome sequencing projects) to describe\\n\\ngene (and\\n\\nprotein) functions and interactions. Functional genomics focuses on the dynamic aspects such as gene\\n\\ntranscription,\\n\\ntranslation, and\\n\\nprotein–protein interactions, as opposed to the static aspects of the genomic information such as\\n\\nDNA sequence or structures. Functional genomics attempts to answer questions about the function of DNA at the levels of genes, RNA transcripts, and protein products. A key characteristic of functional genomics studies is their genome-wide approach to these questions, generally involving high-throughput methods rather than a more traditional \"gene-by-gene\" approach.\\n\\nA major branch of genomics is still concerned with sequencing the genomes of various organisms, but the knowledge of full genomes has created the possibility for the field of functional genomics, mainly concerned with patterns of gene expression during various conditions. The most important tools here are microarrays and bioinformatics.\\n\\nStructural genomics[edit]\\n\\nMain article: \\n\\nStructural genomics\\n\\n3-dimensional structure of every protein encoded by a given\\n\\ngenome.\\n\\n[72]\\n\\n[73]\\n\\nexperimental and modeling approaches. The principal difference between structural genomics and\\n\\ntraditional structural prediction is that structural genomics attempts to determine the structure of every protein encoded by the genome, rather than focusing on one particular protein. With full-genome sequences available, structure prediction can be done more quickly through a combination of experimental and modeling approaches, especially because the availability of large numbers of sequenced genomes and previously solved protein structures allow scientists to model protein structure on the structures of previously solved homologs. Structural genomics involves taking a large number of approaches to structure determination, including experimental methods using genomic sequences or modeling-based approaches based on sequence or\\n\\nstructural homology to a protein of known structure or based on chemical and physical principles for a protein with no homology to any known structure. As opposed to traditional\\n\\nstructural biology, the determination of a\\n\\nprotein structure through a structural genomics effort often (but not always) comes before anything is known regarding the protein function. This raises new challenges in\\n\\nstructural bioinformatics, i.e. determining protein function from its\\n\\n3D structure.\\n\\n[74]\\n\\nEpigenomics[edit]\\n\\nMain article: \\n\\nEpigenomics\\n\\nepigenetic modifications on the genetic material of a cell, known as the\\n\\nepigenome.\\n\\n[75]\\n\\nDNA methylation and\\n\\nhistone modification.\\n\\n[76]\\n\\ndifferentiation/development\\n\\n[77]\\n\\ntumorigenesis.\\n\\n[75]\\n\\n[78]\\n\\nMetagenomics[edit]\\n\\nMain article: \\n\\nMetagenomics', metadata={'source': 'https://en.wikipedia.org/wiki/Genomics'}),\n",
              " Document(page_content='[41]\\n\\n[41]\\n\\n1000 Genomes Project, which announced the sequencing of 1,092 genomes in October 2012.\\n\\n[42]\\n\\nbioinformatics resources from a large international collaboration.\\n\\n[43]\\n\\n[44]\\n\\nThe \"omics\" revolution[edit]\\n\\ngenome,\\n\\ntranscriptome,\\n\\nproteome, and\\n\\nmetabolome (\\n\\nlipidome)\\n\\nMain articles: \\n\\nOmics and\\n\\nHuman proteome project\\n\\nneologism\\n\\nproteomics or\\n\\nmetabolomics. The related suffix\\n\\ngenome,\\n\\nproteome or\\n\\nmetabolome respectively. The suffix\\n\\nJonathan Eisen, among others\\n\\n[45]\\n\\n[46]\\n\\n[47]\\n\\nsymbioses, for example, researchers which were once limited to the study of a single gene product can now simultaneously compare the total complement of several types of biological molecules.\\n\\n[48]\\n\\n[49]\\n\\nGenome analysis[edit]\\n\\nMain article: \\n\\nGenome project\\n\\nAfter an organism has been selected, genome projects involve three components: the sequencing of DNA, the assembly of that sequence to create a representation of the original chromosome, and the annotation and analysis of that representation.[9]\\n\\nBGI or\\n\\nDOE JGI). Third, the genome sequence is annotated at several levels: DNA, protein, gene pathways, or comparatively.\\n\\nSequencing[edit]\\n\\nMain article: \\n\\nDNA Sequencing\\n\\nJoint Genome Institute which sequence dozens of terabases a year, to local molecular biology core facilities)  which contain research laboratories with the costly instrumentation and technical support necessary. As sequencing technology continues to improve, however, a new generation of effective fast turnaround benchtop sequencers has come within reach of the average academic laboratory.\\n\\n[50]\\n\\n[51]\\n\\n[9]\\n\\nShotgun sequencing[edit]\\n\\nMain article: \\n\\nShotgun sequencing\\n\\n[52]\\n\\nshotgun. Since gel electrophoresis sequencing can only be used for fairly short sequences (100 to 1000 base pairs), longer DNA sequences must be broken into random small segments which are then sequenced to obtain\\n\\n[52]\\n\\n[53]\\n\\nnucleotide is represented in the reconstructed sequence; the average number of reads by which a genome is over-sampled is referred to as\\n\\ncoverage.\\n\\n[54]\\n\\nSanger method\\', which is based on the selective incorporation of chain-terminating\\n\\ndideoxynucleotides by\\n\\nDNA polymerase during\\n\\nin vitro\\n\\nDNA replication.\\n\\n[23]\\n\\n[55]\\n\\nhigh-throughput sequencing methods, especially for large-scale, automated\\n\\ngenome analyses. However, the Sanger method remains in wide use, primarily for smaller-scale projects and for obtaining especially long contiguous DNA sequence reads (>500 nucleotides).\\n\\n[56]\\n\\nprimer, a\\n\\nDNA polymerase, normal deoxynucleosidetriphosphates (dNTPs), and modified nucleotides (dideoxyNTPs) that terminate DNA strand elongation. These chain-terminating nucleotides lack a 3\\'-\\n\\nOH group required for the formation of a\\n\\nphosphodiester bond between two nucleotides, causing DNA polymerase to cease extension of DNA when a ddNTP is incorporated. The ddNTPs may be radioactively or\\n\\nfluorescently labelled for detection in\\n\\nDNA sequencers.\\n\\n[9]\\n\\n[57]\\n\\nHigh-throughput sequencing[edit]\\n\\nSee also:', metadata={'source': 'https://en.wikipedia.org/wiki/Genomics'}),\n",
              " Document(page_content='Retrieved from \"\\n\\nhttps://en.wikipedia.org/w/index.php?title=Genomics&oldid=1168149468\"\\n\\nCategory:\\n\\nGenomics\\n\\nHidden categories: \\n\\nPages containing links to subscription-only content\\n\\nArticles with short description\\n\\nShort description is different from Wikidata\\n\\nArticles containing potentially dated statements from October 2011\\n\\nAll articles containing potentially dated statements\\n\\nCS1 maint: DOI inactive as of August 2023\\n\\nWebarchive template wayback links\\n\\nArticles with BNF identifiers\\n\\nArticles with BNFdata identifiers\\n\\nArticles with J9U identifiers\\n\\nArticles with LCCN identifiers\\n\\nArticles with LNB identifiers\\n\\nArticles with NKC identifiers', metadata={'source': 'https://en.wikipedia.org/wiki/Genomics'}),\n",
              " Document(page_content='Shibata T (October 2012). \"Cancer genomics and pathology: all together now\". Pathology International. 62 (10): 647–659. doi:10.1111/j.1440-1827.2012.02855.x. PMID\\xa023005591. S2CID\\xa027886018.\\n\\nRoychowdhury S, Chinnaiyan AM (2016). \"Translating cancer genomes and transcriptomes for precision oncology\". CA. 66 (1): 75–88. doi:10.3322/caac.21329. PMC\\xa04713245. PMID\\xa026528881.\\n\\nGladyshev VN, Zhang Y (2013). \"Chapter 16 Comparative Genomics Analysis of the Metallomes\". In Banci L (ed.). Metallomics and the Cell. Metal Ions in Life Sciences. Vol.\\xa012. Springer. doi:10.1007/978-94-007-5561-10_16 (inactive 1 August 2023). ISBN\\xa0978-94-007-5560-4.{{cite book}}:  CS1 maint: DOI inactive as of August 2023 (link) electronic-book ISBN\\xa0978-94-007-5561-1 ISSN\\xa01559-0836 electronic-ISSN\\xa01868-0402\\n\\nExternal links[edit]\\n\\nAnnual Review of Genomics and Human Genetics Archived 2009-01-18 at the Wayback Machine\\n\\nBMC Genomics: A BMC journal on Genomics\\n\\nGenomics journal\\n\\nGenomics.org: An openfree genomics portal.\\n\\nNHGRI: US government\\'s genome institute\\n\\nJCVI Comprehensive Microbial Resource\\n\\nKoreaGenome.org: The first Korean Genome published and the sequence is available freely.\\n\\nGenomicsNetwork: Looks at the development and use of the science and technologies of genomics.\\n\\nInstitute for Genome Sciences: Genomics research.\\n\\nMIT OpenCourseWare HST.512 Genomic Medicine A free, self-study course in genomic medicine. Resources include audio lectures and selected lecture notes.\\n\\nENCODE threads explorer  Machine learning approaches to genomics. Nature (journal)\\n\\nGlobal map of genomics laboratories\\n\\nGenomics: Scitable by nature education\\n\\nLearn All About Genetics Online\\n\\nv\\n\\nt\\n\\ne\\n\\nOmics\\n\\nGenomics\\n\\nCognitive genomics\\nComputational genomics\\nComparative genomics\\nFunctional genomics\\nGenome project\\nHuman Genome Project\\nMetagenomics\\nHuman Microbiome Project\\nPangenomics\\nPersonal genomics\\nPopulation genomics\\nSocial genomics\\nStructural genomics\\n\\nBioinformatics\\n\\nBiochip\\nCheminformatics\\nChemogenomics\\nConnectomics\\nHuman Connectome Project\\nEpigenomics\\nHuman Epigenome Project\\nGlycomics\\nImmunomics\\nLipidomics\\nMetabolomics\\nMicrobiomics\\nNutrigenomics\\nPaleopolyploidy\\nPharmacogenetics\\nPharmacogenomics\\nSystems biology\\nToxicogenomics\\nTranscriptomics\\n\\nStructural biology\\n\\nProteomics\\nHuman proteome project\\nCall-map proteomics\\nStructure-based drug design\\nExpression proteomics\\n\\n2-D electrophoresis\\nMass spectrometer\\nElectrospray ionization\\nMatrix-assisted laser desorption ionization\\nMatrix-assisted laser desorption ionization-time of flight mass spectrometer\\nMicrofluidic-based tools\\nIsotope affinity tags\\nChromosome conformation capture\\n\\nDNA Data Bank of Japan (JP)\\nEuropean Molecular Biology Laboratory (EU)\\nNational Institutes of Health (USA)\\nWellcome Sanger Institute (UK)\\n\\nList \\n Category\\n\\nv\\n\\nt\\n\\ne\\n\\nGenetics\\n\\nIntroduction\\nOutline\\nHistory\\nTimeline\\nIndex\\nGlossary\\n\\nChromosome\\nDNA\\nRNA\\nGenome\\nHeredity\\nNucleotide\\nMutation\\nGenetic variation\\nAllele\\nAmino acid', metadata={'source': 'https://en.wikipedia.org/wiki/Genomics'})]"
            ]
          },
          "metadata": {},
          "execution_count": 74
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from langchain.chat_models import ChatOpenAI\n",
        "from langchain.llms import OpenAI\n",
        "from langchain.chains.question_answering import load_qa_chain"
      ],
      "metadata": {
        "id": "sQbHiM2i3cJ5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# model_name = \"text-davinci-003\"\n",
        "# model_name = \"gpt-3.5-turbo\"\n",
        "model_name = \"gpt-3.5-turbo\"\n",
        "llm = ChatOpenAI(model_name=model_name)\n",
        "\n",
        "chain = load_qa_chain(llm, chain_type=\"stuff\")\n",
        "\n",
        "def get_answer(query):\n",
        "  similar_docs = get_similiar_docs(query)\n",
        "  answer = chain.run(input_documents=similar_docs, question=query)\n",
        "  return answer\n"
      ],
      "metadata": {
        "id": "2r4_wqTUTeCg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "query= 'types of genomes'"
      ],
      "metadata": {
        "id": "EhwP5VbQ27sm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ans = get_answer(query)\n",
        "ans"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 192
        },
        "id": "Lq79Cit54GOS",
        "outputId": "30dbb575-121b-4b58-8435-11e9fc4f7402"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'There are several types of genomes, including:\\n\\n1. Nuclear Genome: This is the genome found in the nucleus of cells and contains the majority of an organism\\'s genetic material.\\n\\n2. Mitochondrial Genome: Mitochondria have their own separate genome, known as the mitochondrial genome. It is much smaller than the nuclear genome and is responsible for encoding some of the proteins and RNAs necessary for mitochondrial function.\\n\\n3. Chloroplast Genome: Found in plants and algae, the chloroplast genome is responsible for encoding proteins and RNAs necessary for photosynthesis.\\n\\n4. Viral Genome: Viruses have their own genomes, which can be composed of DNA or RNA. These genomes can be single-stranded or double-stranded, linear or circular, and can vary greatly in size.\\n\\n5. Bacterial Genome: Bacteria have their own genomes, which are typically circular DNA molecules. These genomes contain all the genetic information necessary for bacterial survival and reproduction.\\n\\nIt\\'s important to note that the term \"genome\" can also refer to the complete set of genetic material of an individual or a species, regardless of the specific type of genome.'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 95
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "VmkCWe-z4GJ1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "y0ZO8l7G6DFG"
      }
    }
  ]
}